--- a/net/minecraft/advancements/Advancement.java
+++ b/net/minecraft/advancements/Advancement.java
@@ -62,7 +62,7 @@ public record Advancement(
         final AdvancementRequirements requirements,
         final boolean sendsTelemetryEvent
     ) {
-        this(parent, display, rewards, Map.copyOf(criteria), requirements, sendsTelemetryEvent, display.map(Advancement::decorateName));
+        this(parent, display, rewards, criteria, requirements, sendsTelemetryEvent, display.map(Advancement::decorateName));
     }
 
     private static DataResult<Advancement> validate(final Advancement advancement) {
