--- a/paper-api/build.gradle.kts
+++ b/paper-api/build.gradle.kts
@@ -5,7 +_,7 @@
     `java-library`
     `maven-publish`
     idea
-    id("io.papermc.paperweight.paper-checkstyle")
+    //id("io.papermc.paperweight.paper-checkstyle")
 }
 
 java {
@@ -17,10 +_,12 @@
     JavadocTag("apiNote", "a", "API Note:"),
 )
 
+/*
 paperCheckstyle {
     customJavadocTags = projectCustomJavadocTags
     directoriesToSkipFile = layout.projectDirectory.file(".checkstyle/ignored_directories.txt")
 }
+*/
 
 val annotationsVersion = "26.1.0"
 val adventureVersion = "5.2.0"
@@ -103,10 +_,10 @@
     testRuntimeOnly("org.junit.platform:junit-platform-launcher:6.0.3")
 
     // checkstyle
-    checkstyle(project(":paper-checkstyle"))
+    //checkstyle(project(":paper-checkstyle"))
 }
 
-val generatedDir: java.nio.file.Path = layout.projectDirectory.dir("src/generated/java").asFile.toPath()
+val generatedDir: java.nio.file.Path = layout.projectDirectory.dir("../paper-api/src/generated/java").asFile.toPath()
 idea {
     module {
         generatedSourceDirs.add(generatedDir.toFile())
@@ -116,6 +_,18 @@
     main {
         java {
             srcDir(generatedDir)
+            srcDir(file("../paper-api/src/main/java"))
+        }
+        resources {
+            srcDir(file("../paper-api/src/main/resources"))
+        }
+    }
+    test {
+        java {
+            srcDir(file("../paper-api/src/test/java"))
+        }
+        resources {
+            srcDir(file("../paper-api/src/test/resources"))
         }
     }
 }
