--- a/net/minecraft/world/entity/monster/warden/Warden.java
+++ b/net/minecraft/world/entity/monster/warden/Warden.java
@@ -248,7 +_,7 @@
     public void tick() {
         if (this.level() instanceof ServerLevel serverLevel) {
             VibrationSystem.Ticker.tick(serverLevel, this.vibrationData, this.vibrationUser);
-            if (this.isPersistenceRequired() || this.requiresCustomPersistence()) {
+            if (this.gale$eventDriven_isPersistent) { // Gale - Event-driven - Mob.isPersistent
                 WardenAi.setDigCooldown(this);
             }
         }
