--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -489,6 +_,8 @@
         this.bukkitPickUpLoot = true;
         this.maxHealthCache = this.getMaxHealth();
         // CraftBukkit end
+        this.gale$eventDriven_isAlive_update(); // Gale - Event-driven - LivingEntity.isAlive
+        this.gale$eventDriven_affectsSpawningSelector_update(); // Gale - Event-driven - Player.affectsSpawningSelector
     }
 
     @Override
