--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -634,7 +_,7 @@
         savedDataStorage.set(io.papermc.paper.world.saveddata.PaperWorldPDC.TYPE, loadedWorldData.pdc() == null ? io.papermc.paper.world.saveddata.PaperWorldPDC.TYPE.constructor().get() : loadedWorldData.pdc());
         final GameRules gameRules = new GameRules(server.getWorldData().enabledFeatures(), savedDataStorage.computeIfAbsent(net.minecraft.world.level.gamerules.GameRuleMap.TYPE));
         this.gameRules = gameRules;
-        super(levelData, dimension, server.registryAccess(), levelStem.type(), false, isDebug, biomeZoomSeed, server.getMaxChainedNeighborUpdates(), loadedWorldData.bukkitName(), gen, biomeProvider, env, spigotConfig -> server.paperConfigurations.createWorldConfig(io.papermc.paper.configuration.PaperConfigurations.createWorldContextMap(server.storageSource.getDimensionPath(dimension), dimension.identifier(), spigotConfig, server.registryAccess(), gameRules)), executor); // Paper - create paper world configs // Paper - Anti-Xray - Pass executor
+        super(levelData, dimension, server.registryAccess(), levelStem.type(), false, isDebug, biomeZoomSeed, server.getMaxChainedNeighborUpdates(), loadedWorldData.bukkitName(), gen, biomeProvider, env, spigotConfig -> server.paperConfigurations.createWorldConfig(io.papermc.paper.configuration.PaperConfigurations.createWorldContextMap(server.storageSource.getDimensionPath(dimension), dimension.identifier(), spigotConfig, server.registryAccess(), gameRules)), spigotConfig -> null/*server.galeConfigurations.createWorldConfig(io.papermc.paper.configuration.PaperConfigurations.createWorldContextMap(server.storageSource.getDimensionPath(dimension), dimension.identifier(), spigotConfig, server.registryAccess(), gameRules))*/, executor); // Paper - create paper world configs // Paper - Anti-Xray - Pass executor // Gale - Configuration - Create world config // TODO Gale configuration temp disabled
         this.weatherData = savedDataStorage.computeIfAbsent(WeatherData.TYPE);
         this.weatherData.setLevel(this);
         this.typeKey = typeKey;
@@ -797,9 +_,12 @@
     public void tick(final BooleanSupplier haveTime) {
         ProfilerFiller profiler = Profiler.get();
         this.handlingTick = true;
+        if (org.galemc.gale.async.SimulationFlag.REAL) { // Gale - Speculative execution based on next tick simulation - Simulate next tick
         this.environmentAttributes().invalidateTickCache();
+        } // Gale - Speculative execution based on next tick simulation - Simulate next tick
         TickRateManager tickRateManager = this.tickRateManager();
         boolean runs = tickRateManager.runsNormally();
+        if (org.galemc.gale.async.SimulationFlag.REAL) { // Gale - Speculative execution based on next tick simulation - Simulate next tick
         if (runs) {
             profiler.push("world border");
             this.getWorldBorder().tick();
@@ -838,6 +_,7 @@
         if (runs) {
             this.tickTime();
         }
+        } // Gale - Speculative execution based on next tick simulation - Simulate next tick
 
         profiler.push("tickPending");
         if (!this.isDebug() && runs) {
@@ -856,14 +_,17 @@
 
         profiler.popPush("chunkSource");
         this.getChunkSource().tick(haveTime, true);
+        if (org.galemc.gale.async.SimulationFlag.REAL) { // Gale - Speculative execution based on next tick simulation - Simulate next tick
         profiler.popPush("blockEvents");
         if (runs) {
             this.runBlockEvents();
         }
 
+        } // Gale - Speculative execution based on next tick simulation - Simulate next tick
         this.handlingTick = false;
         profiler.pop();
         boolean isActive = !paperConfig().unsupportedSettings.disableWorldTickingWhenEmpty || this.chunkSource.hasActiveTickets(); // CraftBukkit - this prevents entity cleanup, other issues on servers with no players // Paper - restore this
+        if (org.galemc.gale.async.SimulationFlag.REAL) { // Gale - Speculative execution based on next tick simulation - Simulate next tick
         if (isActive) {
             this.resetEmptyTime();
         }
@@ -872,7 +_,8 @@
             this.emptyTime++;
         }
 
-        if (this.emptyTime < 300) {
+        } // Gale - Speculative execution based on next tick simulation - Simulate next tick
+        if (org.galemc.gale.async.SimulationFlag.REAL ? this.emptyTime < 300 : isActive || this.emptyTime < 299) { // Gale - Speculative execution based on next tick simulation - Simulate next tick
             profiler.push("entities");
             if (this.dragonFight != null && runs) {
                 profiler.push("dragonFight");
@@ -889,6 +_,7 @@
                                 profiler.push("checkDespawn");
                                 entity.checkDespawn();
                                 profiler.pop();
+                                if (!org.galemc.gale.async.SimulationFlag.REAL) return; // Gale - Speculative execution based on next tick simulation - Simulate next tick
                                 if (true) { // Paper - rewrite chunk system
                                     Entity vehicle = entity.getVehicle();
                                     if (vehicle != null) {
@@ -912,9 +_,10 @@
             this.tickBlockEntities();
             profiler.pop();
             } // Paper - option to disable ticking
-            this.spigotConfig.currentPrimedTnt = 0; // Spigot
+            if (!org.galemc.gale.async.SimulationFlag.REAL) this.spigotConfig.currentPrimedTnt = 0; // Spigot // Gale - Speculative execution based on next tick simulation - Simulate next tick
         }
 
+        if (org.galemc.gale.async.SimulationFlag.REAL) { // Gale - Speculative execution based on next tick simulation - Simulate next tick
         profiler.push("entityManagement");
         // Paper - rewrite chunk system
         profiler.pop();
@@ -928,6 +_,7 @@
 
         this.debugSynchronizers.tick(this.server.debugSubscribers());
         profiler.pop();
+        } // Gale - Speculative execution based on next tick simulation - Simulate next tick
     }
 
     @Override
@@ -1395,7 +_,7 @@
             // Paper end - log detailed entity tick information
         entity.setOldPosAndRot();
         ProfilerFiller profiler = Profiler.get();
-        entity.tickCount++;
+        entity.setTickCount(entity.tickCount + 1); // Gale - Event-driven - Cat.canRemoveWhenFarAway, Ocelot.canRemoveWhenFarAway
         entity.totalEntityAge++; // Paper - age-like counter for all entities
         profiler.push(entity.typeHolder()::getRegisteredName);
         profiler.incrementCounter("tickNonPassenger");
@@ -1421,9 +_,9 @@
     private void tickPassenger(final Entity vehicle, final Entity entity, final boolean isActive) { // Paper - EAR 2
         if (entity.isRemoved() || entity.getVehicle() != vehicle) {
             entity.stopRiding();
-        } else if (entity instanceof Player || this.entityTickList.contains(entity)) {
+        } else if (entity instanceof Player || entity.gale$eventDriven_isInTickList) { // Gale - Event-driven - Entity.isInTickList
             entity.setOldPosAndRot();
-            entity.tickCount++;
+            entity.setTickCount(entity.tickCount + 1); // Gale - Event-driven - Cat.canRemoveWhenFarAway, Ocelot.canRemoveWhenFarAway
             entity.totalEntityAge++; // Paper - age-like counter for all entities
             ProfilerFiller profiler = Profiler.get();
             profiler.push(entity.typeHolder()::getRegisteredName);
@@ -1716,6 +_,13 @@
 
     @Override
     public void destroyBlockProgress(final int id, final BlockPos blockPos, final int progress) {
+        // Gale start - Local code optimization - ServerLevel.destroyBlockProgress() (allocate zero or one block destruction packets)
+        var players = this.server.getPlayerList().getPlayers();
+        if (players.isEmpty()) {
+            return;
+        }
+        ClientboundBlockDestructionPacket blockDestructionPacket = new ClientboundBlockDestructionPacket(id, blockPos, progress);
+        // Gale end - Local code optimization - ServerLevel.destroyBlockProgress() (allocate zero or one block destruction packets)
         // CraftBukkit start
         Player breakerPlayer = null;
         Entity entity = this.getEntity(id);
@@ -1732,7 +_,7 @@
                 .callEvent();
         }
         // Paper end - Add BlockBreakProgressUpdateEvent
-        for (ServerPlayer player : this.server.getPlayerList().getPlayers()) {
+        for (ServerPlayer player : players) { // Gale - Local code optimization - ServerLevel.destroyBlockProgress() (allocate zero or one block destruction packets)
             if (player.level() == this && player.getId() != id) {
                 double xd = blockPos.getX() - player.getX();
                 double yd = blockPos.getY() - player.getY();
@@ -1743,7 +_,7 @@
                         continue;
                     }
                     // CraftBukkit end
-                    player.connection.send(new ClientboundBlockDestructionPacket(id, blockPos, progress));
+                    player.connection.send(blockDestructionPacket); // Gale - Local code optimization - ServerLevel.destroyBlockProgress() (allocate zero or one block destruction packets)
                 }
             }
         }
@@ -2386,26 +_,28 @@
 
     @Override
     public void updatePOIOnBlockStateChange(final BlockPos pos, final BlockState oldState, final BlockState newState) {
-        Optional<Holder<PoiType>> oldType = PoiTypes.forState(oldState);
-        Optional<Holder<PoiType>> newType = PoiTypes.forState(newState);
-        if (!Objects.equals(oldType, newType)) {
+        // Gale start - Do less work - PoiTypes.forState() (skip Optional creation)
+        Holder<PoiType> oldType = PoiTypes.gale$forStateNullable(oldState);
+        Holder<PoiType> newType = PoiTypes.gale$forStateNullable(newState);
+        if (oldType != newType) {
             BlockPos immutable = pos.immutable();
-            oldType.ifPresent(poiType -> this.getServer().execute(() -> {
+            if (oldType != null) this.getServer().execute(() -> {
                 this.getPoiManager().remove(immutable);
                 this.debugSynchronizers.dropPoi(immutable);
-            }));
-            newType.ifPresent(poiType -> this.getServer().execute(() -> {
+            });
+            if (newType != null) this.getServer().execute(() -> {
                 // Paper start - Remove stale POIs
-                if (oldType.isEmpty() && this.getPoiManager().exists(immutable, _ -> true)) {
+                if (oldType == null && this.getPoiManager().exists(immutable, _ -> true)) {
                     this.getPoiManager().remove(immutable);
                 }
                 // Paper end - Remove stale POIs
-                PoiRecord record = this.getPoiManager().add(immutable, (Holder<PoiType>)poiType);
+                PoiRecord record = this.getPoiManager().add(immutable, newType);
                 if (record != null) {
                     this.debugSynchronizers.registerPoi(record);
                 }
-            }));
+            });
         }
+        // Gale end - Do less work - PoiTypes.forState() (skip Optional creation)
     }
 
     public PoiManager getPoiManager() {
@@ -2856,6 +_,7 @@
             // ServerLevel.this.getChunkSource().addEntity(entity); // Paper - ignore and warn about illegal addEntity calls instead of crashing server; moved down below valid=true
             if (entity instanceof ServerPlayer player) {
                 ServerLevel.this.players.add(player);
+                player.gale$eventDriven_trackedByLevel_set(ServerLevel.this); // Gale - Event-driven - Player.affectsSpawningSelector
                 if (player.isReceivingWaypoints()) {
                     ServerLevel.this.getWaypointManager().addPlayer(player);
                 }
@@ -2934,6 +_,7 @@
             ServerLevel.this.getChunkSource().removeEntity(entity);
             if (entity instanceof ServerPlayer player) {
                 ServerLevel.this.players.remove(player);
+                player.gale$eventDriven_trackedByLevel_set(null); // Gale - Event-driven - Player.affectsSpawningSelector
                 ServerLevel.this.getWaypointManager().removePlayer(player);
                 ServerLevel.this.updateSleepingPlayerList();
             }
