--- a/net/minecraft/world/level/dimension/end/EnderDragonFight.java
+++ b/net/minecraft/world/level/dimension/end/EnderDragonFight.java
@@ -465,7 +_,7 @@
             .registryAccess()
             .lookup(Registries.CONFIGURED_FEATURE)
             .flatMap(registry -> registry.get(EndFeatures.END_GATEWAY_DELAYED))
-            .ifPresent(endGateway -> endGateway.value().place(this.level, this.level.getChunkSource().getGenerator(), RandomSource.create(), pos));
+            .ifPresent(endGateway -> endGateway.value().place(this.level, this.level.getChunkSource().getGenerator(), this.level.getRandom(), pos)); // Gale - Reduce RandomSource instances
     }
 
     public void spawnExitPortal(final boolean activated) {
