--- a/net/minecraft/world/entity/ai/targeting/TargetingConditions.java
+++ b/net/minecraft/world/entity/ai/targeting/TargetingConditions.java
@@ -79,9 +_,17 @@
             }
 
             if (this.range > 0.0) {
-                double modifier = this.testInvisible ? target.getVisibilityPercent(targeter) : 1.0;
-                double visibilityDistance = Math.max(this.range * modifier, 2.0);
                 double distanceToSqr = targeter.distanceToSqr(target.getX(), target.getY(), target.getZ());
+                // Gale start - Check targeting range before getting visibility
+                double followRangeRaw = this.range;
+
+                if (distanceToSqr > followRangeRaw * followRangeRaw) {
+                    return false;
+                }
+
+                double modifier = this.testInvisible ? target.getVisibilityPercent(targeter) : 1.0;
+                double visibilityDistance = Math.max(followRangeRaw * modifier, 2.0);
+                // Gale end - Check targeting range before getting visibility
                 if (distanceToSqr > visibilityDistance * visibilityDistance) {
                     return false;
                 }
