--- a/net/minecraft/world/level/EntityBasedExplosionDamageCalculator.java
+++ b/net/minecraft/world/level/EntityBasedExplosionDamageCalculator.java
@@ -17,8 +_,20 @@
     public Optional<Float> getBlockExplosionResistance(
         final Explosion explosion, final BlockGetter level, final BlockPos pos, final BlockState block, final FluidState fluid
     ) {
-        return super.getBlockExplosionResistance(explosion, level, pos, block, fluid)
-            .map(resistance -> this.source.getBlockExplosionResistance(explosion, level, pos, block, fluid, resistance));
+        // Gale start - Local code optimization - EntityBasedExplosionDamageCalculator#getBlockExplosionResistance (reduce allocations)
+        Optional<Float> optionalBlastResistance = super.getBlockExplosionResistance(explosion, level, pos, block, fluid);
+
+        if (optionalBlastResistance.isPresent()) {
+            float resistance = optionalBlastResistance.get();
+            float effectiveExplosionResistance = this.source.getBlockExplosionResistance(explosion, level, pos, block, fluid, resistance);
+
+            if (effectiveExplosionResistance != resistance) {
+                return Optional.of(effectiveExplosionResistance);
+            }
+        }
+
+        return optionalBlastResistance;
+        // Gale end - Local code optimization - EntityBasedExplosionDamageCalculator#getBlockExplosionResistance (reduce allocations)
     }
 
     @Override
