--- a/net/minecraft/world/level/LocalMobCapCalculator.java
+++ b/net/minecraft/world/level/LocalMobCapCalculator.java
@@ -42,14 +_,14 @@
     }
 
     private static class MobCounts {
-        private final Object2IntMap<MobCategory> counts = new Object2IntOpenHashMap<>(MobCategory.values().length);
+        private final int[] counts = new int[MobCategory.values().length]; // Gale - Store mob counts in an array
 
         public void add(final MobCategory category) {
-            this.counts.computeInt(category, (k, count) -> count == null ? 1 : count + 1);
+            this.counts[category.ordinal()]++; // Gale - Store mob counts in an array
         }
 
         public boolean canSpawn(final MobCategory category) {
-            return this.counts.getOrDefault(category, 0) < category.getMaxInstancesPerChunk();
+            return this.counts[category.ordinal()] < category.getMaxInstancesPerChunk(); // Gale - Store mob counts in an array
         }
     }
 }
