--- a/net/minecraft/world/level/NaturalSpawner.java
+++ b/net/minecraft/world/level/NaturalSpawner.java
@@ -86,7 +_,7 @@
         Object2IntOpenHashMap<MobCategory> mobCounts = new Object2IntOpenHashMap<>();
 
         for (Entity entity : entities) {
-            if (!(entity instanceof Mob mob && (mob.isPersistenceRequired() || mob.requiresCustomPersistence()))) {
+            if (!(entity instanceof Mob mob && mob.gale$eventDriven_isPersistent)) { // Gale - Event-driven - Mob.isPersistent
                 MobCategory category = entity.getType().getCategory();
                 if (category != MobCategory.MISC) {
                     // Paper start - Only count natural spawns
@@ -477,7 +_,11 @@
     public static boolean isInNetherFortressBounds(
         final BlockPos pos, final ServerLevel level, final MobCategory category, final StructureManager structureManager
     ) {
-        if (category == MobCategory.MONSTER && level.getBlockState(pos.below()).is(Blocks.NETHER_BRICKS)) {
+        // Gale start - Don't load chunks synchronously - Nether fortress mob spawn
+        if (category == MobCategory.MONSTER) {
+            BlockState blockState = level.getBlockStateIfLoaded(pos.below());
+            if (blockState == null || !blockState.is(Blocks.NETHER_BRICKS)) return false;
+            // Gale end - Don't load chunks synchronously - Nether fortress mob spawn
             Structure fortress = structureManager.registryAccess().lookupOrThrow(Registries.STRUCTURE).getValue(BuiltinStructures.FORTRESS);
             return fortress != null && structureManager.getStructureAt(pos, fortress).isValid();
         } else {
