--- a/net/minecraft/world/entity/ai/behavior/GateBehavior.java
+++ b/net/minecraft/world/entity/ai/behavior/GateBehavior.java
@@ -85,7 +_,16 @@
             }
         }
         // Paper end - Perf: Remove streams from hot code
-        if (this.behaviors.stream().noneMatch(behavior -> behavior.getStatus() == Behavior.Status.RUNNING)) {
+        // Gale start - Leaf #0134: Remove stream in GateBehavior
+        boolean hasRunningTask = false;
+        for (final BehaviorControl<? super E> behavior : this.behaviors) {
+            if (behavior.getStatus() == Behavior.Status.RUNNING) {
+                hasRunningTask = true;
+                break;
+            }
+        }
+        if (!hasRunningTask) {
+        // Gale end - Leaf #0134: Remove stream in GateBehavior
             this.doStop(level, body, timestamp);
         }
     }
