--- a/net/minecraft/world/entity/ai/behavior/InteractWithDoor.java
+++ b/net/minecraft/world/entity/ai/behavior/InteractWithDoor.java
@@ -53,7 +_,7 @@
                             Node toNode = path.getNextNode();
                             BlockPos fromPos = fromNode.asBlockPos();
                             BlockState fromState = level.getBlockState(fromPos);
-                            if (fromState.is(BlockTags.MOB_INTERACTABLE_DOORS, s -> s.getBlock() instanceof DoorBlock)) {
+                            if (fromState.isAny(org.dreeam.leaf.util.BlockMasks.MOB_INTERACTABLE_DOORS_TAG_AND_DOOR_CLASS)) { // Gale - Pre-compute - Common BlockState predicates - Apply
                                 DoorBlock fromBlock = (DoorBlock)fromState.getBlock();
                                 if (!fromBlock.isOpen(fromState)) {
                                     // CraftBukkit start - entities opening doors
@@ -70,7 +_,7 @@
 
                             BlockPos toPos = toNode.asBlockPos();
                             BlockState toState = level.getBlockState(toPos);
-                            if (toState.is(BlockTags.MOB_INTERACTABLE_DOORS, s -> s.getBlock() instanceof DoorBlock)) {
+                            if (toState.isAny(org.dreeam.leaf.util.BlockMasks.MOB_INTERACTABLE_DOORS_TAG_AND_DOOR_CLASS)) { // Gale - Pre-compute - Common BlockState predicates - Apply
                                 DoorBlock door = (DoorBlock)toState.getBlock();
                                 if (!door.isOpen(toState)) {
                                     // CraftBukkit start - entities opening doors
@@ -117,7 +_,7 @@
                     iterator.remove();
                 } else {
                     BlockState state = level.getBlockState(doorPos);
-                    if (!state.is(BlockTags.MOB_INTERACTABLE_DOORS, s -> s.getBlock() instanceof DoorBlock)) {
+                    if (!state.isAny(org.dreeam.leaf.util.BlockMasks.MOB_INTERACTABLE_DOORS_TAG_AND_DOOR_CLASS)) { // Gale - Pre-compute - Common BlockState predicates - Apply
                         iterator.remove();
                     } else {
                         DoorBlock block = (DoorBlock)state.getBlock();
