--- a/net/minecraft/world/entity/ai/behavior/SleepInBed.java
+++ b/net/minecraft/world/entity/ai/behavior/SleepInBed.java
@@ -59,7 +_,7 @@
 
         BlockState blockState = level.getBlockStateIfLoaded(target.pos()); // Paper - Prevent sync chunk loads when villagers try to find beds
         if (blockState == null) return false; // Paper - Prevent sync chunk loads when villagers try to find beds
-        return target.pos().closerToCenterThan(body.position(), 2.0) && blockState.is(BlockTags.BEDS) && !blockState.getValue(BedBlock.OCCUPIED);
+        return target.pos().closerToCenterThan(body.position(), 2.0) && blockState.isUnoccupiedBed(); // Gale - Pre-compute - Common BlockState predicates - Apply
     }
 
     @Override
