--- a/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -490,6 +_,13 @@
         private boolean emptyCollisionShape;
         private boolean emptyConstantCollisionShape;
         private VoxelShape constantCollisionShape;
+        // Gale start - Pre-compute - Common BlockState predicates
+        private int predicateFlags;
+        public boolean isAny(int predicateMask) {
+            return (this.predicateFlags & predicateMask) != 0;
+        }
+        public void initPredicateFlags() { this.predicateFlags = org.dreeam.leaf.util.BlockMasks.init(this.asState()); }
+        // Gale end - Pre-compute - Common BlockState predicates
 
         private static void initCaches(final VoxelShape shape, final boolean neighbours) {
             ((ca.spottedleaf.moonrise.patches.collisions.shape.CollisionVoxelShape)shape).moonrise$isFullBlock();
@@ -668,6 +_,13 @@
         public Holder<Block> typeHolder() {
             return this.getBlock().builtInRegistryHolder();
         }
+
+        // Gale start - Faster implementation for subclass - BlockBehaviour.BlockStateBase.is(Block)
+        @Override
+        public boolean is(final Block rawType) {
+            return this.owner == rawType;
+        }
+        // Gale end - Faster implementation for subclass - BlockBehaviour.BlockStateBase.is(Block)
 
         @Deprecated
         public boolean blocksMotion() {
