--- a/src/test/java/org/bukkit/support/suite/LegacyTestSuite.java
+++ b/src/test/java/org/bukkit/support/suite/LegacyTestSuite.java
@@ -9,7 +_,7 @@
 @Suite(failIfNoTests = false)
 @SuiteDisplayName("Test suite for legacy tests")
 @IncludeTags("Legacy")
-@SelectPackages({"org.bukkit", "io.papermc.paper", "com.destroystokyo.paper"})
+@SelectPackages({"org.bukkit", "io.papermc.paper", "com.destroystokyo.paper", "org.galemc", "org.dreeam"}) // Gale - Unit tests
 @ConfigurationParameter(key = "TestSuite", value = "Legacy")
 public class LegacyTestSuite {
 }
