--- a/src/test/java/org/bukkit/support/suite/SlowTestSuite.java
+++ b/src/test/java/org/bukkit/support/suite/SlowTestSuite.java
@@ -9,7 +_,7 @@
 @Suite(failIfNoTests = false)
 @SuiteDisplayName("Test suite for slow tests, which don't need to run every time")
 @IncludeTags("Slow")
-@SelectPackages({"org.bukkit", "io.papermc.paper", "com.destroystokyo.paper"})
+@SelectPackages({"org.bukkit", "io.papermc.paper", "com.destroystokyo.paper", "org.galemc", "org.dreeam"}) // Gale - Unit tests
 @ConfigurationParameter(key = "TestSuite", value = "Slow")
 public class SlowTestSuite {
 }
