--- a/net/minecraft/world/entity/monster/Creeper.java
+++ b/net/minecraft/world/entity/monster/Creeper.java
@@ -250,6 +_,7 @@
             if (!event.isCancelled()) {
             // CraftBukkit end
             this.dead = true;
+            this.gale$eventDriven_isAlive_update(); // Gale - Event-driven - LivingEntity.isAlive
             level.explode(this, this.getX(), this.getY(), this.getZ(), event.getRadius(), event.getFire(), Level.ExplosionInteraction.MOB); // CraftBukkit // Paper - fix DamageSource API (revert to vanilla, no, just no, don't change this)
             this.spawnLingeringCloud();
             this.triggerOnDeathMobEffects(level, Entity.RemovalReason.KILLED);
