--- a/net/minecraft/world/entity/monster/EnderMan.java
+++ b/net/minecraft/world/entity/monster/EnderMan.java
@@ -298,11 +_,15 @@
     private boolean teleport(final double x, final double y, final double z) {
         BlockPos.MutableBlockPos pos = new BlockPos.MutableBlockPos(x, y, z);
 
-        while (pos.getY() > this.level().getMinY() && !this.level().getBlockState(pos).blocksMotion()) {
+        // Gale start - Skip enderman teleport if requires main thread chunk load
+        net.minecraft.world.level.chunk.LevelChunk chunk = this.level().getChunkIfLoaded(pos);
+        if (chunk == null) return false;
+        while (pos.getY() > this.level().getMinY() && !chunk.getBlockState(pos).blocksMotion()) {
+            // Gale end - Skip enderman teleport if requires main thread chunk load
             pos.move(Direction.DOWN);
         }
 
-        BlockState blockState = this.level().getBlockState(pos);
+        BlockState blockState = chunk.getBlockState(pos); // Gale - Airplane - reduce enderman teleport chunk lookups
         boolean couldStandOn = blockState.blocksMotion();
         boolean isWet = blockState.getFluidState().is(FluidTags.WATER);
         if (couldStandOn && !isWet) {
