--- a/net/minecraft/world/entity/monster/EnderMan.java
+++ b/net/minecraft/world/entity/monster/EnderMan.java
@@ -73,7 +_,7 @@
     );
     private static final int DELAY_BETWEEN_CREEPY_STARE_SOUND = 400;
     private static final int MIN_DEAGGRESSION_TIME = 600;
-    private static final EntityDataAccessor<Optional<BlockState>> DATA_CARRY_STATE = SynchedEntityData.defineId(
+    private static final EntityDataAccessor<Optional<BlockState>> DATA_CARRY_STATE = SynchedEntityData.defineId( // Gale - Event-driven - Mob.requiresCustomPersistence - TODO Watch for new set calls
         EnderMan.class, EntityDataSerializers.OPTIONAL_BLOCK_STATE
     );
     private static final EntityDataAccessor<Boolean> DATA_CREEPY = SynchedEntityData.defineId(EnderMan.class, EntityDataSerializers.BOOLEAN);
@@ -87,6 +_,7 @@
     public EnderMan(final EntityType<? extends EnderMan> type, final Level level) {
         super(type, level);
         this.setPathfindingMalus(PathType.WATER, -1.0F);
+        this.gale$eventDriven_requiresCustomPersistence_update(); // Gale - Event-driven - Mob.requiresCustomPersistence
     }
 
     @Override
@@ -298,11 +_,15 @@
     private boolean teleport(final double x, final double y, final double z) {
         BlockPos.MutableBlockPos pos = new BlockPos.MutableBlockPos(x, y, z);
 
-        while (pos.getY() > this.level().getMinY() && !this.level().getBlockState(pos).blocksMotion()) {
+        // Gale start - Don't load chunks synchronously - Enderman teleport
+        net.minecraft.world.level.chunk.LevelChunk chunk = this.level().getChunkIfLoaded(pos);
+        if (chunk == null) return false;
+        while (pos.getY() > this.level().getMinY() && !chunk.getBlockState(pos).blocksMotion()) {
+            // Gale end - Don't load chunks synchronously - Enderman teleport
             pos.move(Direction.DOWN);
         }
 
-        BlockState blockState = this.level().getBlockState(pos);
+        BlockState blockState = chunk.getBlockState(pos); // Gale - Airplane - reduce enderman teleport chunk lookups
         boolean couldStandOn = blockState.blocksMotion();
         boolean isWet = blockState.getFluidState().is(FluidTags.WATER);
         if (couldStandOn && !isWet) {
@@ -363,6 +_,7 @@
 
     public void setCarriedBlock(final @Nullable BlockState carryingBlock) {
         this.entityData.set(DATA_CARRY_STATE, Optional.ofNullable(carryingBlock));
+        this.gale$eventDriven_requiresCustomPersistence_update(); // Gale - Event-driven - Mob.requiresCustomPersistence
     }
 
     public @Nullable BlockState getCarriedBlock() {
@@ -429,8 +_,10 @@
     // Paper end
 
     @Override
-    public boolean requiresCustomPersistence() {
-        return super.requiresCustomPersistence() || this.getCarriedBlock() != null;
+    // Gale start - Event-driven - Mob.requiresCustomPersistence
+    public boolean gale$eventDriven_requiresCustomPersistence_compute() {
+        return super.gale$eventDriven_requiresCustomPersistence_compute() || this.getCarriedBlock() != null;
+        // Gale end - Event-driven - Mob.requiresCustomPersistence
     }
 
     private static class EndermanFreezeWhenLookedAt extends Goal {
