--- a/net/minecraft/world/entity/npc/villager/Villager.java
+++ b/net/minecraft/world/entity/npc/villager/Villager.java
@@ -843,8 +_,14 @@
     }
 
     private int countFoodPointsInInventory() {
+        // Gale start - avoid stream in villager food check
         SimpleContainer inventory = this.getInventory();
-        return FOOD_POINTS.entrySet().stream().mapToInt(entry -> inventory.countItem(entry.getKey()) * entry.getValue()).sum();
+        int total = 0;
+        for (Map.Entry<Item, Integer> entry : FOOD_POINTS.entrySet()) {
+            total += inventory.countItem(entry.getKey()) * entry.getValue();
+        }
+        return total;
+        // Gale end - avoid stream in villager food check
     }
 
     public boolean hasFarmSeeds() {
