--- a/net/minecraft/world/entity/ai/sensing/SecondaryPoiSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/SecondaryPoiSensor.java
@@ -22,16 +_,22 @@
 
     @Override
     protected void doTick(final ServerLevel level, final Villager body) {
+        // Gale start - Do less work - SecondaryPoiSensor.doTick() (skip secondary POI sensor if absent)
+        var secondaryPoi = body.getVillagerData().profession().value().secondaryPoi();
+        if (secondaryPoi.isEmpty()) {
+            body.getBrain().eraseMemory(MemoryModuleType.SECONDARY_JOB_SITE);
+            return;
+        }
+        // Gale end - Do less work - SecondaryPoiSensor.doTick() (skip secondary POI sensor if absent)
         ResourceKey<Level> dimensionType = level.dimension();
         BlockPos center = body.blockPosition();
         List<GlobalPos> jobSites = Lists.newArrayList();
-        int horizontalSearch = 4;
 
         for (int x = -4; x <= 4; x++) {
             for (int y = -2; y <= 2; y++) {
                 for (int z = -4; z <= 4; z++) {
                     BlockPos testPos = center.offset(x, y, z);
-                    if (body.getVillagerData().profession().value().secondaryPoi().contains(level.getBlockState(testPos).getBlock())) {
+                    if (secondaryPoi.contains(level.getBlockState(testPos).getBlock())) { // Gale - Do less work - SecondaryPoiSensor.doTick() (skip secondary POI sensor if absent)
                         jobSites.add(GlobalPos.of(dimensionType, testPos));
                     }
                 }
