--- a/net/minecraft/world/entity/ai/sensing/Sensing.java
+++ b/net/minecraft/world/entity/ai/sensing/Sensing.java
@@ -9,8 +_,10 @@
 
 public class Sensing {
     private final Mob mob;
-    private final IntSet seen = new IntOpenHashSet();
-    private final IntSet unseen = new IntOpenHashSet();
+    // Gale start - Collections - Sensing
+    private final IntSet seen = new IntOpenHashSet(2);
+    private final IntSet unseen = new IntOpenHashSet(2);
+    // Gale end - Collections - Sensing
 
     public Sensing(final Mob mob) {
         this.mob = mob;
