--- a/src/main/java/io/papermc/paper/configuration/PaperConfigurations.java
+++ b/src/main/java/io/papermc/paper/configuration/PaperConfigurations.java
@@ -335,6 +_,7 @@
             this.initializeWorldDefaultsConfiguration(server.registryAccess());
             for (ServerLevel level : server.getAllLevels()) {
                 this.createWorldConfig(createWorldContextMap(level), reloader(this.worldConfigClass, level.paperConfig()));
+                level.gale$checkDespawn_postUpdateSpawnRanges(); // Gale - Optimize Mob.checkDespawn()
             }
         } catch (Exception ex) {
             throw new RuntimeException("Could not reload paper configuration files", ex);
@@ -482,7 +_,7 @@
     }
 
     // Symlinks are not correctly checked in createDirectories
-    static void createDirectoriesSymlinkAware(Path path) throws IOException {
+    public static void createDirectoriesSymlinkAware(Path path) throws IOException { // Gale - Configuration - Make public
         if (!Files.isDirectory(path)) {
             Files.createDirectories(path);
         }
