--- a/src/main/java/io/papermc/paper/configuration/PaperConfigurations.java
+++ b/src/main/java/io/papermc/paper/configuration/PaperConfigurations.java
@@ -335,6 +_,11 @@
             this.initializeWorldDefaultsConfiguration(server.registryAccess());
             for (ServerLevel level : server.getAllLevels()) {
                 this.createWorldConfig(createWorldContextMap(level), reloader(this.worldConfigClass, level.paperConfig()));
+                // Gale start - Event-driven - Mob.despawnRanges, Mob.despawnRangeShape
+                for (net.minecraft.world.entity.Entity entity : level.getEntities().getAll()) {
+                    if (entity instanceof net.minecraft.world.entity.Mob mob) mob.gale$eventDriven_despawnRanges_and_despawnRangeShape_update();
+                }
+                // Gale end - Event-driven - Mob.despawnRanges, Mob.despawnRangeShape
             }
         } catch (Exception ex) {
             throw new RuntimeException("Could not reload paper configuration files", ex);
@@ -482,7 +_,7 @@
     }
 
     // Symlinks are not correctly checked in createDirectories
-    static void createDirectoriesSymlinkAware(Path path) throws IOException {
+    public static void createDirectoriesSymlinkAware(Path path) throws IOException { // Gale - Configuration - Make public
         if (!Files.isDirectory(path)) {
             Files.createDirectories(path);
         }
