--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -328,6 +_,12 @@
     private static final int MAX_SIGN_LINE_LENGTH = Integer.getInteger("Paper.maxSignLength", 80); // Paper - Limit client sign length
     private final io.papermc.paper.event.packet.ClientTickEndEvent tickEndEvent; // Paper - add client tick end event
     public final io.papermc.paper.connection.PaperPlayerGameConnection playerGameConnection; // Paper
+    // Gale start - Collections - ChunkMap.TrackedEntity.seenBy
+    public static final org.galemc.gale.util.collection.packed.UsedPackedIndices gale$USED_PACKED_INDICES = new org.galemc.gale.util.collection.packed.UsedPackedIndices();
+    public final int gale$packedIndex;
+    public final int gale$packedWordIndex;
+    public final long gale$packedWordMask;
+    // Gale end - Collections - ChunkMap.TrackedEntity.seenBy
 
     public ServerGamePacketListenerImpl(final MinecraftServer server, final Connection connection, final ServerPlayer player, final CommonListenerCookie cookie) {
         super(server, connection, cookie);
@@ -342,6 +_,12 @@
         this.chatMessageChain = new FutureChain(server.chatExecutor); // CraftBukkit - async chat
         this.tickEndEvent = new io.papermc.paper.event.packet.ClientTickEndEvent(player.getBukkitEntity()); // Paper - add client tick end event
         this.playerGameConnection = new io.papermc.paper.connection.PaperPlayerGameConnection(this); // Paper
+        // Gale start - Collections - ChunkMap.TrackedEntity.seenBy
+        int gale$packedIndex = gale$USED_PACKED_INDICES.claimUnused();
+        this.gale$packedIndex = gale$packedIndex;
+        this.gale$packedWordIndex = org.galemc.gale.util.collection.packed.PackedBooleanArray.getWordIndex(gale$packedIndex);
+        this.gale$packedWordMask = org.galemc.gale.util.collection.packed.PackedBooleanArray.getWordMask(gale$packedIndex);
+        // Gale end - Collections - ChunkMap.TrackedEntity.seenBy
     }
 
     // Paper start - configuration phase API
