--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -514,15 +_,17 @@
         // CraftBukkit start
         // this.broadcastAll(new ClientboundPlayerInfoRemovePacket(List.of(player.getUUID())));
         ClientboundPlayerInfoRemovePacket packet = new ClientboundPlayerInfoRemovePacket(List.of(player.getUUID()));
+        org.bukkit.craftbukkit.entity.CraftPlayer craftPlayer = player.getBukkitEntity(); // Gale - Store canSee in a fast-access format
         for (int i = 0; i < this.players.size(); i++) {
             ServerPlayer otherPlayer = this.players.get(i);
 
-            if (otherPlayer.getBukkitEntity().canSee(player.getBukkitEntity())) {
+            if (otherPlayer.getBukkitEntity().canSee(craftPlayer)) { // Gale - Store canSee in a fast-access format
                 otherPlayer.connection.send(packet);
             } else {
                 otherPlayer.getBukkitEntity().onEntityRemove(player);
             }
         }
+        craftPlayer.resetCanSee(); // Gale - Store canSee in a fast-access format
         // This removes the scoreboard (and player reference) for the specific player in the manager
         this.cserver.getScoreboardManager().removePlayer(player.getBukkitEntity());
         // CraftBukkit end
