--- a/net/minecraft/world/entity/EntitySelector.java
+++ b/net/minecraft/world/entity/EntitySelector.java
@@ -47,9 +_,8 @@
         Team.CollisionRule ownCollisionRule = ownTeam == null ? Team.CollisionRule.ALWAYS : ownTeam.getCollisionRule();
         return ownCollisionRule == Team.CollisionRule.NEVER
             ? Predicates.alwaysFalse()
-            : NO_SPECTATORS.and(
-                input -> {
-                    if (!input.isPushable() || !input.canCollideWithBukkit(entity) || !entity.canCollideWithBukkit(input)) { // CraftBukkit - collidable API // Paper - Climbing should not bypass cramming gamerule
+            : input -> { // Gale - Do less work - EntitySelector.pushableBy()
+                    if (input.isSpectator() /*|| !pushedEntity.isPushable()*/ || !input.canCollideWithBukkit(entity) || !entity.canCollideWithBukkit(input)) { // CraftBukkit - collidable API // Paper - Climbing should not bypass cramming gamerule // Gale - Do less work - EntitySelector.pushableBy()
                         return false;
                     }
 
@@ -69,8 +_,7 @@
                     } else {
                         return false;
                     }
-                }
-            );
+                }; // Gale - Do less work - EntitySelector.pushableBy()
     }
 
     public static Predicate<Entity> notRiding(final Entity entity) {
