--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -561,13 +_,14 @@
     }
 
     protected void tryAddFrost() {
-        if (!this.getBlockStateOnLegacy().isAir()) {
+//        if (!this.getBlockStateOnLegacy().isAir()) { // Gale - Sort tryAddFrost checks
             int ticksFrozen = this.getTicksFrozen();
             if (ticksFrozen > 0) {
                 AttributeInstance speed = this.getAttribute(Attributes.MOVEMENT_SPEED);
                 if (speed == null) {
                     return;
                 }
+        if (!this.getBlockStateOnLegacy().isAir()) { // Gale - Sort tryAddFrost checks
 
                 float slowAmount = -0.05F * this.getPercentFrozen();
                 speed.addTransientModifier(new AttributeModifier(SPEED_MODIFIER_POWDER_SNOW_ID, slowAmount, AttributeModifier.Operation.ADD_VALUE));
@@ -946,6 +_,7 @@
 
     protected void tickEffects() {
         if (this.level() instanceof ServerLevel serverLevel) {
+            if (!this.activeEffects.isEmpty()) { // Gale - Don't iterate over empty tick effect collections
             Iterator<Holder<MobEffect>> iterator = this.activeEffects.keySet().iterator();
 
             this.tickingEffects = true; // CraftBukkit
@@ -970,6 +_,7 @@
             }
             // CraftBukkit start
             this.tickingEffects = false;
+            } // Gale - Don't iterate over empty tick effect collections
             Runnable task;
             while ((task = this.deferredEffectUpdates.poll()) != null) {
                 task.run();
@@ -2160,7 +_,7 @@
         } else {
             BlockPos ladderCheckPos = this.blockPosition();
             BlockState state = this.getInBlockState();
-            if (this.isFallFlying() && state.is(BlockTags.CAN_GLIDE_THROUGH)) {
+            if (this.isFallFlying() && state.isAny(org.dreeam.leaf.util.BlockMasks.CAN_GLIDE_THROUGH_TAG)) { // Gale - Pre-compute block state predicates - Apply
                 return false;
             } else if (state.is(BlockTags.CLIMBABLE)) {
                 this.lastClimbablePos = Optional.of(ladderCheckPos);
@@ -3793,7 +_,13 @@
         if (this.level() instanceof ServerLevel serverLevel) {
             profiler.push("freezing");
             if ((!this.isInPowderSnow || !this.canFreeze()) && !this.freezeLocked) { // Paper - Freeze Tick Lock API
-                this.setTicksFrozen(Math.max(0, this.getTicksFrozen() - 2));
+                // Gale start - Only update frozen ticks if changed
+                int ticksFrozen = this.getTicksFrozen();
+                int newTicksFrozen = Math.max(0, ticksFrozen - 2);
+                if (ticksFrozen != newTicksFrozen) {
+                    this.setTicksFrozen(newTicksFrozen);
+                }
+                // Gale end - Only update frozen ticks if changed
             }
 
             this.removeFrost();
