--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -189,7 +_,7 @@
     protected static final int LIVING_ENTITY_FLAG_OFF_HAND = 2;
     public static final int LIVING_ENTITY_FLAG_SPIN_ATTACK = 4;
     protected static final EntityDataAccessor<Byte> DATA_LIVING_ENTITY_FLAGS = SynchedEntityData.defineId(LivingEntity.class, EntityDataSerializers.BYTE);
-    public static final EntityDataAccessor<Float> DATA_HEALTH_ID = SynchedEntityData.defineId(LivingEntity.class, EntityDataSerializers.FLOAT);
+    public static final EntityDataAccessor<Float> DATA_HEALTH_ID = SynchedEntityData.defineId(LivingEntity.class, EntityDataSerializers.FLOAT); // Gale - Event-driven - LivingEntity.isAlive - TODO Watch for new set calls
     private static final EntityDataAccessor<List<ParticleOptions>> DATA_EFFECT_PARTICLES = SynchedEntityData.defineId(
         LivingEntity.class, EntityDataSerializers.PARTICLES
     );
@@ -254,7 +_,7 @@
     public final ElytraAnimationState elytraAnimationState = new ElytraAnimationState(this);
     public @Nullable EntityReference<Player> lastHurtByPlayer;
     public int lastHurtByPlayerMemoryTime;
-    protected boolean dead;
+    protected boolean dead; // Gale - Event-driven - LivingEntity.isAlive - TODO Watch for new writes
     protected int noActionTime;
     public float lastHurt;
     protected boolean jumping;
@@ -291,7 +_,7 @@
     private final EnumMap<EquipmentSlot, Reference2ObjectMap<Enchantment, Set<EnchantmentLocationBasedEffect>>> activeLocationDependentEnchantments = new EnumMap<>(
         EquipmentSlot.class
     );
-    protected final EntityEquipment equipment;
+    protected final EntityEquipment equipment; // Gale - Event-driven - Mob.requiresCustomPersistence - TODO Watch for new writes
     private Waypoint.Icon locatorBarIcon = new Waypoint.Icon();
     public @Nullable Vec3 currentImpulseImpactPos;
     // CraftBukkit start
@@ -319,6 +_,7 @@
         this.setYRot(this.random.nextFloat() * (float) (Math.PI * 2));
         this.yHeadRot = this.getYRot();
         this.brain = this.makeBrain(Brain.Packed.EMPTY);
+        this.gale$eventDriven_isAlive_updateFor_health(this.entityData.get(DATA_HEALTH_ID)); // Gale - Event-driven - LivingEntity.isAlive
     }
 
     @Override
@@ -561,13 +_,14 @@
     }
 
     protected void tryAddFrost() {
-        if (!this.getBlockStateOnLegacy().isAir()) {
+//        if (!this.getBlockStateOnLegacy().isAir()) { // Gale - Do less work - LivingEntity.tryAddFrost()
             int ticksFrozen = this.getTicksFrozen();
             if (ticksFrozen > 0) {
                 AttributeInstance speed = this.getAttribute(Attributes.MOVEMENT_SPEED);
                 if (speed == null) {
                     return;
                 }
+        if (!this.getBlockStateOnLegacy().isAir()) { // Gale - Do less work - LivingEntity.tryAddFrost()
 
                 float slowAmount = -0.05F * this.getPercentFrozen();
                 speed.addTransientModifier(new AttributeModifier(SPEED_MODIFIER_POWDER_SNOW_ID, slowAmount, AttributeModifier.Operation.ADD_VALUE));
@@ -946,6 +_,7 @@
 
     protected void tickEffects() {
         if (this.level() instanceof ServerLevel serverLevel) {
+            if (!this.activeEffects.isEmpty()) { // Gale - Local code optimization - LivingEntity.tickEffects() (don't iterate over empty effect collection)
             Iterator<Holder<MobEffect>> iterator = this.activeEffects.keySet().iterator();
 
             this.tickingEffects = true; // CraftBukkit
@@ -970,6 +_,7 @@
             }
             // CraftBukkit start
             this.tickingEffects = false;
+            } // Gale - Local code optimization - LivingEntity.tickEffects() (don't iterate over empty effect collection)
             Runnable task;
             while ((task = this.deferredEffectUpdates.poll()) != null) {
                 task.run();
@@ -1416,7 +_,11 @@
             return;
         }
         // CraftBukkit end
-        this.entityData.set(DATA_HEALTH_ID, Mth.clamp(health, 0.0F, this.getMaxHealth()));
+        // Gale start - Event-driven - LivingEntity.isAlive
+        float newHealth = Mth.clamp(health, 0.0F, this.getMaxHealth());
+        this.entityData.set(DATA_HEALTH_ID, newHealth);
+        this.gale$eventDriven_isAlive_updateFor_health(newHealth);
+        // Gale end - Event-driven - LivingEntity.isAlive
     }
 
     public boolean isDeadOrDying() {
@@ -1857,6 +_,7 @@
                 this.deathDropItems = null;
                 if (deathEvent.isCancelled()) {
                     this.dead = false;
+                    this.gale$eventDriven_isAlive_update(); // Gale - Event-driven - LivingEntity.isAlive
                     this.postDeathEventTasks.clear();
                     this.setHealth((float) deathEvent.getReviveHealth());
                     return;
@@ -1881,6 +_,7 @@
 
     protected void handleKillingBlow() {
         this.dead = true;
+        this.gale$eventDriven_isAlive_update(); // Gale - Event-driven - LivingEntity.isAlive
     }
 
     protected void createWitherRose(final @Nullable LivingEntity killer) {
@@ -2159,9 +_,9 @@
         } else {
             BlockPos ladderCheckPos = this.blockPosition();
             BlockState state = this.getInBlockState();
-            if (this.isFallFlying() && state.is(BlockTags.CAN_GLIDE_THROUGH)) {
+            if (this.isFallFlying() && state.isAny(org.dreeam.leaf.util.BlockMasks.CAN_GLIDE_THROUGH_TAG)) { // Gale - Pre-compute - Common BlockState predicates - Apply
                 return false;
-            } else if (state.is(BlockTags.CLIMBABLE)) {
+            } else if (state.isAny(org.dreeam.leaf.util.BlockMasks.CLIMBABLE_TAG)) { // Gale - Pre-compute - Common BlockState predicates - Apply
                 this.lastClimbablePos = Optional.of(ladderCheckPos);
                 return true;
             } else if (state.getBlock() instanceof TrapDoorBlock && this.trapdoorUsableAsLadder(ladderCheckPos, state)) {
@@ -2184,8 +_,30 @@
 
     @Override
     public boolean isAlive() {
-        return !this.isRemoved() && this.getHealth() > 0.0F && !this.dead; // Paper - Check this.dead
-    }
+        // Gale start - Event-driven - LivingEntity.isAlive
+        return this.gale$eventDriven_isAlive;
+    }
+
+    public boolean gale$eventDriven_isAlive_compute(float health) {
+        return !this.isRemoved() && health > 0.0F && !this.dead; // Paper - Check this.dead
+    }
+
+    protected boolean gale$eventDriven_isAlive; // Gale - Event-driven - Player.affectsSpawningSelector - TODO Watch for new writes
+
+    protected void gale$eventDriven_isAlive_update() {
+        this.gale$eventDriven_isAlive_updateFor_health(this.getHealth());
+    }
+
+    public void gale$eventDriven_isAlive_updateFor_health(float health) {
+        this.gale$eventDriven_isAlive = this.gale$eventDriven_isAlive_compute(health);
+    }
+
+    @Override
+    public void unsetRemoved() {
+        super.unsetRemoved();
+        this.gale$eventDriven_isAlive_update();
+    }
+    // Gale end - Event-driven - LivingEntity.isAlive
 
     public boolean isLookingAtMe(
         final LivingEntity target,
@@ -3792,7 +_,13 @@
         if (this.level() instanceof ServerLevel serverLevel) {
             profiler.push("freezing");
             if ((!this.isInPowderSnow || !this.canFreeze()) && !this.freezeLocked) { // Paper - Freeze Tick Lock API
-                this.setTicksFrozen(Math.max(0, this.getTicksFrozen() - 2));
+                // Gale start - Do less work - Check ticks frozen difference only once
+                int currentTicksFrozen = this.getTicksFrozen();
+                int newTicksFrozen = Math.max(0, currentTicksFrozen - 2);
+                if (newTicksFrozen != currentTicksFrozen) {
+                    this.entityData.set(DATA_TICKS_FROZEN, newTicksFrozen, true);
+                }
+                // Gale end - Do less work - Check ticks frozen difference only once
             }
 
             this.removeFrost();
