--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -146,15 +_,15 @@
     private final Sensing sensing;
     private DropChances dropChances = DropChances.DEFAULT;
     private boolean canPickUpLoot = false;
-    public boolean persistenceRequired = false;
+    public boolean persistenceRequired = false; // Gale - Event-driven - Mob.isPersistent, Piglin.canRemoveWhenFarAway - TODO Watch for new writes
     private final Map<PathType, Float> pathfindingMalus = Maps.newEnumMap(PathType.class);
     public Optional<ResourceKey<LootTable>> lootTable = Optional.empty();
     public long lootTableSeed;
-    private Leashable.@Nullable LeashData leashData;
+    private Leashable.@Nullable LeashData leashData; // Gale - Event-driven - Mob.isLeashed - TODO Watch for new writes
     private BlockPos homePosition = BlockPos.ZERO;
     private int homeRadius = -1;
     public boolean aware = true; // CraftBukkit
-    public net.kyori.adventure.util.TriState despawnInPeacefulOverride = net.kyori.adventure.util.TriState.NOT_SET; // Paper - allow changing despawnInPeaceful
+    public net.kyori.adventure.util.TriState despawnInPeacefulOverride = net.kyori.adventure.util.TriState.NOT_SET; // Paper - allow changing despawnInPeaceful // Gale - Event-driven - Mob.shouldDespawnInPeaceful - TODO Watch for new writes
 
     protected Mob(final EntityType<? extends Mob> type, final Level level) {
         super(type, level);
@@ -169,6 +_,9 @@
         if (level instanceof ServerLevel) {
             this.registerGoals();
         }
+        this.gale$eventDriven_checkDespawnPlayerTrackingMode_update(); // Gale - Event-driven - Mob.checkDespawnPlayerTrackingMode
+        this.gale$eventDriven_despawnRanges_and_despawnRangeShape_update(); // Gale - Event-driven - Mob.despawnRanges, Mob.despawnRangeShape
+        this.gale$eventDriven_mustDespawnBecauseOfPeacefulDifficulty_update(); // Gale - Event-driven - Mob.mustDespawnBecauseOfPeacefulDifficulty
     }
 
     protected void registerGoals() {
@@ -469,6 +_,8 @@
         boolean persistenceRequired = input.getBooleanOr("PersistenceRequired", false);
         if (isLevelAtLeast(input, 1) || persistenceRequired) {
             this.persistenceRequired = persistenceRequired;
+            this.gale$eventDriven_isPersistent_update(); // Gale - Event-driven - Mob.isPersistent
+           if (this instanceof net.minecraft.world.entity.monster.piglin.Piglin piglin) piglin.gale$eventDriven_canRemoveWhenFarAway_update(); // Gale - Event-driven - Piglin.canRemoveWhenFarAway
         }
         // CraftBukkit end
         this.dropChances = input.read("drop_chances", DropChances.CODEC).orElse(DropChances.DEFAULT);
@@ -485,6 +_,7 @@
         this.aware = input.getBooleanOr("Bukkit.Aware", true); // CraftBukkit
         // Paper start - allow changing despawnInPeaceful
         this.despawnInPeacefulOverride = readDespawnInPeacefulOverride(input);
+        this.gale$eventDriven_gale$shouldDespawnInPeaceful_update(); // Gale - Event-driven - Mob.shouldDespawnInPeaceful
     }
     public static net.kyori.adventure.util.TriState readDespawnInPeacefulOverride(ValueInput input) {
         return input.read("Paper.DespawnInPeacefulOverride", io.papermc.paper.util.PaperCodecs.TRI_STATE_CODEC).orElse(net.kyori.adventure.util.TriState.NOT_SET);
@@ -653,6 +_,8 @@
             ItemStack toEquip = slot.limit(itemStack);
             this.setItemSlotAndDropWhenKilled(slot, toEquip);
             this.persistenceRequired = true;
+            this.gale$eventDriven_isPersistent_update(); // Gale - Event-driven - Mob.isPersistent
+            if (this instanceof net.minecraft.world.entity.monster.piglin.Piglin piglin) piglin.gale$eventDriven_canRemoveWhenFarAway_update(); // Gale - Event-driven - Piglin.canRemoveWhenFarAway
             return toEquip;
         } else {
             return ItemStack.EMPTY;
@@ -780,38 +_,196 @@
         return true;
     }
 
-    public boolean requiresCustomPersistence() {
+    private io.papermc.paper.configuration.WorldConfiguration.Entities.Spawning.DespawnRangePair gale$eventDriven_despawnRanges;
+    private io.papermc.paper.configuration.type.DespawnRange.Shape gale$eventDriven_despawnRangeShape;
+
+    // Gale start - Event-driven - Mob.despawnRanges, Mob.despawnRangeShape
+    public void gale$eventDriven_despawnRanges_and_despawnRangeShape_update() {
+        io.papermc.paper.configuration.WorldConfiguration config = this.level().paperConfig();
+        io.papermc.paper.configuration.WorldConfiguration.Entities.Spawning.DespawnRangePair newDespawnRanges = config.entities.spawning.despawnRanges.get(this.getType().getCategory());
+        io.papermc.paper.configuration.type.DespawnRange.Shape newDespawnRangeShape = config.entities.spawning.despawnRangeShape;
+        if (!newDespawnRanges.effectivelyEquals(this.gale$eventDriven_despawnRanges) || !newDespawnRangeShape.equals(this.gale$eventDriven_despawnRangeShape)) {
+            this.gale$eventDriven_despawnRanges = newDespawnRanges;
+            this.gale$eventDriven_despawnRangeShape = newDespawnRangeShape;
+        }
+    }
+    // Gale end - Event-driven - Mob.despawnRanges, Mob.despawnRangeShape
+
+    // Gale start - Optimize Mob.checkDespawn - Pre-check for removeWhenFarAway()
+    /**
+     * @return Whether it's currently possible for {@link #removeWhenFarAway(double)} to return true
+     * for some distance. In other words, this method returns false if {@link #removeWhenFarAway(double)} currently
+     * returns false for any given distance, and this method returns true if {@link #removeWhenFarAway(double)}
+     * may currently return true or false.
+     */
+    public boolean gale$canRemoveWhenFarAway() { // Gale - Event-driven - Mob.checkDespawnPlayerTrackingMode - TODO Watch for changes in return value
+        return this.removeWhenFarAway(-1);
+    }
+
+    public void gale$eventDriven_canRemoveWhenFarAway_postUpdate() {
+        this.gale$eventDriven_checkDespawnPlayerTrackingMode_update(); // Gale - Event-driven - Mob.checkDespawnPlayerTrackingMode
+    }
+
+    public boolean gale$eventDriven_requiresCustomPersistence_compute() { // Gale - Event-driven - Mob.requiresCustomPersistence
         return this.isPassenger() || this.isLeashed();
     }
+    // Gale end - Optimize Mob.checkDespawn - Pre-check for removeWhenFarAway()
+
+    // Gale start - Event-driven - Mob.requiresCustomPersistence
+    public boolean requiresCustomPersistence() {
+        return this.gale$eventDriven_requiresCustomPersistence;
+    }
+
+    private boolean gale$eventDriven_requiresCustomPersistence; // Gale - Event-driven - Mob.isPersistent - TODO Watch for new writes
+
+    protected void gale$eventDriven_requiresCustomPersistence_update() {
+        boolean newValue = this.gale$eventDriven_requiresCustomPersistence_compute();
+        if (this.gale$eventDriven_requiresCustomPersistence != newValue) {
+            this.gale$eventDriven_requiresCustomPersistence = newValue;
+            this.gale$eventDriven_isPersistent_update();
+        }
+    }
+
+    @Override
+    public void setVehicle(@Nullable Entity vehicle) {
+        super.setVehicle(vehicle);
+        this.gale$eventDriven_requiresCustomPersistence_update();
+    }
+    // Gale end - Event-driven - Mob.requiresCustomPersistence
 
     // Paper start - allow changing despawnInPeaceful
-    public final boolean shouldDespawnInPeaceful() {
+    public final boolean gale$eventDriven_shouldDespawnInPeaceful_compute() { // Gale - Event-driven - Mob.shouldDespawnInPeaceful
         return this.despawnInPeacefulOverride.toBooleanOrElse(!this.getType().isTypeAllowedInPeaceful());
     }
     // Paper end - allow changing despawnInPeaceful
 
+    // Gale start - Event-driven - Mob.shouldDespawnInPeaceful
+    public final boolean shouldDespawnInPeaceful() {
+        return this.gale$eventDriven_shouldDespawnInPeaceful; // Gale - Event-driven - Mob.shouldDespawnInPeaceful
+    }
+
+    private boolean gale$eventDriven_shouldDespawnInPeaceful;
+
+    public void gale$eventDriven_gale$shouldDespawnInPeaceful_update() {
+        boolean newValue = this.gale$eventDriven_shouldDespawnInPeaceful_compute();
+        if (this.gale$eventDriven_shouldDespawnInPeaceful != newValue) {
+            this.gale$eventDriven_shouldDespawnInPeaceful = newValue;
+            this.gale$eventDriven_mustDespawnBecauseOfPeacefulDifficulty_update();
+        }
+    }
+    // Gale end - Event-driven - Mob.shouldDespawnInPeaceful
+
+    // Gale start - Event-driven - Mob.mustDespawnBecauseOfPeacefulDifficulty
+    protected boolean gale$eventDriven_mustDespawnBecauseOfPeacefulDifficulty; // Gale - Event-driven - Mob.checkDespawnPlayerTrackingMode - TODO Watch for new writes
+
+    public void gale$eventDriven_mustDespawnBecauseOfPeacefulDifficulty_update() {
+        boolean newValue = this.level().getDifficulty() == Difficulty.PEACEFUL && this.shouldDespawnInPeaceful();
+        if (this.gale$eventDriven_mustDespawnBecauseOfPeacefulDifficulty != newValue) {
+            this.gale$eventDriven_mustDespawnBecauseOfPeacefulDifficulty = newValue;
+            this.gale$eventDriven_checkDespawnPlayerTrackingMode_update(); // Gale - Event-driven - Mob.checkDespawnPlayerTrackingMode
+        }
+    }
+
+    @Override
+    public void setLevel(final Level level) {
+        super.setLevel(level);
+        this.gale$eventDriven_despawnRanges_and_despawnRangeShape_update(); // Gale - Event-driven - Mob.despawnRanges, Mob.despawnRangeShape
+        this.gale$eventDriven_mustDespawnBecauseOfPeacefulDifficulty_update();
+    }
+    // Gale end - Event-driven - Mob.mustDespawnBecauseOfPeacefulDifficulty
+
+    // Gale start - Event-driven - Mob.isLeashed
+    protected boolean gale$eventDriven_isLeashed; // Gale - Event-driven - Mob.requiresCustomPersistence, Ocelot.canRemoveWhenFarAway - TODO Watch for new writes
+    private java.lang.ref.@Nullable WeakReference<net.minecraft.world.entity.Mob> leashDataListenerReference;
+
+    public void gale$eventDriven_isLeashed_update() {
+        boolean newValue = this.leashData != null && this.leashData.leashHolder != null;
+        if (this.gale$eventDriven_isLeashed != newValue) {
+            this.gale$eventDriven_isLeashed = newValue;
+            this.gale$eventDriven_requiresCustomPersistence_update();
+        }
+    }
+
+    @Override
+    public boolean isLeashed() {
+        return this.gale$eventDriven_isLeashed;
+    }
+    // Gale end - Event-driven - Mob.isLeashed
+
+    // Gale start - Event-driven - Mob.isPersistent
+    public boolean gale$eventDriven_isPersistent; // Gale - Event-driven - Mob.checkDespawnPlayerTrackingMode - TODO Watch for new writes
+
+    public void gale$eventDriven_isPersistent_update() {
+        boolean newValue = this.persistenceRequired || this.gale$eventDriven_requiresCustomPersistence;
+        if (this.gale$eventDriven_isPersistent != newValue) {
+            this.gale$eventDriven_isPersistent = newValue;
+            this.gale$eventDriven_checkDespawnPlayerTrackingMode_update(); // Gale - Event-driven - Mob.checkDespawnPlayerTrackingMode
+        }
+    }
+    // Gale end - Event-driven - Mob.isPersistent
+
+    // Gale start - Optimize Mob.checkDespawn
+    /**
+     * Whether this mob needs to track players for {@link #checkDespawn}:
+     * <ul>
+     *     <li>0 = none</li>
+     *     <li>1 = only for the {@link io.papermc.paper.configuration.WorldConfiguration.Entities.Spawning.DespawnRangePair#soft()} distance</li>
+     *     <li>2 = for both {@link io.papermc.paper.configuration.WorldConfiguration.Entities.Spawning.DespawnRangePair#soft()} and
+     *     {@link io.papermc.paper.configuration.WorldConfiguration.Entities.Spawning.DespawnRangePair#hard()} distances</li>
+     * </ul>
+     */
+    public int gale$eventDriven_checkDespawnPlayerTrackingMode;
+
+    public void gale$eventDriven_checkDespawnPlayerTrackingMode_update() {
+        this.gale$eventDriven_checkDespawnPlayerTrackingMode = this.gale$eventDriven_checkDespawnPlayerTrackingMode_compute();
+    }
+
+    private byte gale$eventDriven_checkDespawnPlayerTrackingMode_compute() {
+        if (!this.gale$eventDriven_isInTickList || this.gale$eventDriven_mustDespawnBecauseOfPeacefulDifficulty || this.gale$eventDriven_isPersistent || this.level().gale$eventDriven_affectsSpawningSelectorPlayerCount == 0) {
+            return 0;
+        } else if (!this.gale$canRemoveWhenFarAway()) {
+            return 1;
+        } else {
+            return 2;
+        }
+    }
+
+    @Override
+    public void gale$eventDriven_isInTickList_set(boolean newValue) {
+        boolean oldValue = this.gale$eventDriven_isInTickList;
+        if (oldValue != newValue) {
+            super.gale$eventDriven_isInTickList_set(newValue);
+            this.gale$eventDriven_checkDespawnPlayerTrackingMode_update();
+        }
+    }
+    // Gale end - Optimize Mob.checkDespawn
+
     @Override
     public void checkDespawn() {
-        if (this.level().getDifficulty() == Difficulty.PEACEFUL && this.shouldDespawnInPeaceful()) { // Paper - allow changing despawnInPeaceful
+        if (this.gale$eventDriven_mustDespawnBecauseOfPeacefulDifficulty) { // Paper - allow changing despawnInPeaceful // Gale - Event-driven - Mob.mustDespawnBecauseOfPeacefulDifficulty
             this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
-        } else if (!this.isPersistenceRequired() && !this.requiresCustomPersistence()) {
+        } else if (!this.gale$eventDriven_isPersistent) { // Gale - Event-driven - Mob.isPersistent
+            if (!org.galemc.gale.async.SimulationFlag.REAL) return; // Gale - Speculative execution based on next tick simulation - Simulate next tick
+            if (this.level().gale$eventDriven_affectsSpawningSelectorPlayerCount != 0) { // Gale - Optimize Mob.checkDespawn
             Entity player = this.level().findNearbyPlayer(this, -1.0, EntitySelector.PLAYER_AFFECTS_SPAWNING); // Paper - Affects Spawning API
-            if (player != null) {
                 // Paper start - Configurable despawn distances
-                final io.papermc.paper.configuration.WorldConfiguration.Entities.Spawning.DespawnRangePair despawnRangePair = this.level().paperConfig().entities.spawning.despawnRanges.get(this.getType().getCategory());
-                final io.papermc.paper.configuration.type.DespawnRange.Shape shape = this.level().paperConfig().entities.spawning.despawnRangeShape;
+                // Gale start - Event-driven - Mob.despawnRanges, Mob.despawnRangeShape
+                final io.papermc.paper.configuration.WorldConfiguration.Entities.Spawning.DespawnRangePair despawnRangePair = this.gale$eventDriven_despawnRanges;
+                final io.papermc.paper.configuration.type.DespawnRange.Shape shape = this.gale$eventDriven_despawnRangeShape;
+                // Gale end - Event-driven - Mob.despawnRanges, Mob.despawnRangeShape
                 final double dy = Math.abs(player.getY() - this.getY());
                 final double dySqr = Mth.square(dy);
                 final double dxSqr = Mth.square(player.getX() - this.getX());
                 final double dzSqr = Mth.square(player.getZ() - this.getZ());
                 final double distanceSquared = dxSqr + dzSqr + dySqr;
                 // Despawn if hard/soft limit is exceeded
-                if (despawnRangePair.hard().shouldDespawn(shape, dxSqr, dySqr, dzSqr, dy) && this.removeWhenFarAway(distanceSquared)) {
+                boolean canRemoveWhenFarAway = this.gale$canRemoveWhenFarAway(); // Gale - Optimize Mob.checkDespawn
+                if (canRemoveWhenFarAway && despawnRangePair.hard().shouldDespawn(shape, dxSqr, dySqr, dzSqr, dy) && this.removeWhenFarAway(distanceSquared)) { // Gale - Optimize Mob.checkDespawn
                     this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
                 }
 
-                if (despawnRangePair.soft().shouldDespawn(shape, dxSqr, dySqr, dzSqr, dy)) {
-                    if (this.noActionTime > 600 && this.random.nextInt(800) == 0 && this.removeWhenFarAway(distanceSquared)) {
+                if (despawnRangePair.soft().shouldDespawn(shape, dxSqr, dySqr, dzSqr, dy)) { // Gale - Optimize Mob.checkDespawn
+                    if (canRemoveWhenFarAway && this.noActionTime > 600 && this.random.nextInt(800) == 0 && this.removeWhenFarAway(distanceSquared)) {
                         this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
                     }
                 } else {
@@ -820,6 +_,7 @@
                 }
             }
         } else {
+            if (!org.galemc.gale.async.SimulationFlag.REAL) return; // Gale - Speculative execution based on next tick simulation - Simulate next tick
             this.noActionTime = 0;
         }
     }
@@ -1293,6 +_,7 @@
 
     public void setPersistenceRequired() {
         this.persistenceRequired = true;
+        this.gale$eventDriven_isPersistent_update(); // Gale - Event-driven - Mob.isPersistent
     }
 
     @Override
@@ -1505,7 +_,21 @@
 
     @Override
     public void setLeashData(final Leashable.@Nullable LeashData leashData) {
-        this.leashData = leashData;
+        // Gale start - Event-driven - Mob.isLeashed
+        if (this.leashData != leashData) {
+            if (this.leashData != null && this.leashDataListenerReference != null) {
+                this.leashData.removeLeashHolderListener(this.leashDataListenerReference);
+            }
+            this.leashData = leashData;
+            this.gale$eventDriven_isLeashed_update();
+            if (leashData != null) {
+                if (this.leashDataListenerReference == null) {
+                    this.leashDataListenerReference = new java.lang.ref.WeakReference<>(this);
+                }
+                leashData.addLeashHolderListener(this.leashDataListenerReference);
+            }
+        }
+        // Gale end - Event-driven - Mob.isLeashed
     }
 
     @Override
