--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -146,11 +_,11 @@
     private final Sensing sensing;
     private DropChances dropChances = DropChances.DEFAULT;
     private boolean canPickUpLoot = false;
-    public boolean persistenceRequired = false;
+    public boolean persistenceRequired = false; // Gale - Event-driven - Mob.isPersistent, Piglin.canRemoveWhenFarAway - TODO Watch for new writes
     private final Map<PathType, Float> pathfindingMalus = Maps.newEnumMap(PathType.class);
     public Optional<ResourceKey<LootTable>> lootTable = Optional.empty();
     public long lootTableSeed;
-    private Leashable.@Nullable LeashData leashData;
+    private Leashable.@Nullable LeashData leashData; // Gale - Event-driven - Mob.isLeashed - TODO Watch for new writes
     private BlockPos homePosition = BlockPos.ZERO;
     private int homeRadius = -1;
     public boolean aware = true; // CraftBukkit
@@ -168,6 +_,8 @@
         if (level instanceof ServerLevel) {
             this.registerGoals();
         }
+        this.gale$eventDriven_checkDespawnPlayerTrackingMode_update(); // Gale - Event-driven - Mob.checkDespawnPlayerTrackingMode
+        this.gale$eventDriven_mustDespawnBecauseOfPeacefulDifficulty_update(); // Gale - Event-driven - Mob.mustDespawnBecauseOfPeacefulDifficulty
     }
 
     protected void registerGoals() {
@@ -463,6 +_,8 @@
         boolean persistenceRequired = input.getBooleanOr("PersistenceRequired", false);
         if (isLevelAtLeast(input, 1) || persistenceRequired) {
             this.persistenceRequired = persistenceRequired;
+            this.gale$eventDriven_isPersistent_update(); // Gale - Event-driven - Mob.isPersistent
+           if (this instanceof net.minecraft.world.entity.monster.piglin.Piglin piglin) piglin.gale$eventDriven_canRemoveWhenFarAway_update(); // Gale - Event-driven - Piglin.canRemoveWhenFarAway
         }
         // CraftBukkit end
         this.dropChances = input.read("drop_chances", DropChances.CODEC).orElse(DropChances.DEFAULT);
@@ -641,6 +_,8 @@
             ItemStack toEquip = slot.limit(itemStack);
             this.setItemSlotAndDropWhenKilled(slot, toEquip);
             this.persistenceRequired = true;
+            this.gale$eventDriven_isPersistent_update(); // Gale - Event-driven - Mob.isPersistent
+            if (this instanceof net.minecraft.world.entity.monster.piglin.Piglin piglin) piglin.gale$eventDriven_canRemoveWhenFarAway_update(); // Gale - Event-driven - Piglin.canRemoveWhenFarAway
             return toEquip;
         } else {
             return ItemStack.EMPTY;
@@ -768,43 +_,360 @@
         return true;
     }
 
-    public boolean requiresCustomPersistence() {
+    // Gale start - Optimize Mob.checkDespawn() - Pre-check for removeWhenFarAway()
+    /**
+     * @return Whether it's currently possible for {@link #removeWhenFarAway(double)} to return true
+     * for some distance. In other words, this method returns false if {@link #removeWhenFarAway(double)} currently
+     * returns false for any given distance, and this method returns true if {@link #removeWhenFarAway(double)}
+     * may currently return true or false.
+     */
+    public boolean gale$canRemoveWhenFarAway() { // Gale - Event-driven - Mob.checkDespawnPlayerTrackingMode - TODO Watch for changes in return value
+        return this.removeWhenFarAway(-1);
+    }
+
+    public void gale$eventDriven_canRemoveWhenFarAway_postUpdate() {
+        this.gale$eventDriven_checkDespawnPlayerTrackingMode_update(); // Gale - Event-driven - Mob.checkDespawnPlayerTrackingMode
+    }
+
+    public boolean gale$eventDriven_requiresCustomPersistence_compute() { // Gale - Event-driven - Mob.requiresCustomPersistence
         return this.isPassenger() || this.isLeashed();
     }
+    // Gale end - Optimize Mob.checkDespawn() - Pre-check for removeWhenFarAway()
+
+    // Gale start - Event-driven - Mob.requiresCustomPersistence
+    public boolean requiresCustomPersistence() {
+        return this.gale$eventDriven_requiresCustomPersistence;
+    }
+
+    private boolean gale$eventDriven_requiresCustomPersistence; // Gale - Event-driven - Mob.isPersistent - TODO Watch for new writes
+
+    protected void gale$eventDriven_requiresCustomPersistence_update() {
+        boolean newValue = this.gale$eventDriven_requiresCustomPersistence_compute();
+        if (this.gale$eventDriven_requiresCustomPersistence != newValue) {
+            this.gale$eventDriven_requiresCustomPersistence = newValue;
+            this.gale$eventDriven_isPersistent_update();
+        }
+    }
+
+    @Override
+    public void setVehicle(@Nullable Entity vehicle) {
+        super.setVehicle(vehicle);
+        this.gale$eventDriven_requiresCustomPersistence_update();
+    }
+    // Gale end - Event-driven - Mob.requiresCustomPersistence
+
+    // Gale start - Event-driven - Mob.mustDespawnBecauseOfPeacefulDifficulty
+    protected boolean gale$eventDriven_mustDespawnBecauseOfPeacefulDifficulty; // Gale - Event-driven - Mob.checkDespawnPlayerTrackingMode - TODO Watch for new writes
+
+    public void gale$eventDriven_mustDespawnBecauseOfPeacefulDifficulty_update() {
+        boolean newValue = this.level().getDifficulty() == Difficulty.PEACEFUL && !this.getType().isAllowedInPeaceful();
+        if (this.gale$eventDriven_mustDespawnBecauseOfPeacefulDifficulty != newValue) {
+            this.gale$eventDriven_mustDespawnBecauseOfPeacefulDifficulty = newValue;
+            this.gale$eventDriven_checkDespawnPlayerTrackingMode_update(); // Gale - Event-driven - Mob.checkDespawnPlayerTrackingMode
+        }
+    }
+
+    @Override
+    public void setLevel(final Level level) {
+        super.setLevel(level);
+        this.gale$eventDriven_mustDespawnBecauseOfPeacefulDifficulty_update();
+    }
+    // Gale end - Event-driven - Mob.mustDespawnBecauseOfPeacefulDifficulty
+
+    // Gale start - Event-driven - Mob.isLeashed
+    protected boolean gale$eventDriven_isLeashed; // Gale - Event-driven - Mob.requiresCustomPersistence, Ocelot.canRemoveWhenFarAway - TODO Watch for new writes
+    private java.lang.ref.@Nullable WeakReference<net.minecraft.world.entity.Mob> leashDataListenerReference;
+
+    public void gale$eventDriven_isLeashed_update() {
+        boolean newValue = this.leashData != null && this.leashData.leashHolder != null;
+        if (this.gale$eventDriven_isLeashed != newValue) {
+            this.gale$eventDriven_isLeashed = newValue;
+            this.gale$eventDriven_requiresCustomPersistence_update();
+        }
+    }
+
+    @Override
+    public boolean isLeashed() {
+        return this.gale$eventDriven_isLeashed;
+    }
+    // Gale end - Event-driven - Mob.isLeashed
+
+    // Gale start - Event-driven - Mob.isPersistent
+    public boolean gale$eventDriven_isPersistent; // Gale - Event-driven - Mob.checkDespawnPlayerTrackingMode - TODO Watch for new writes
+
+    public void gale$eventDriven_isPersistent_update() {
+        boolean newValue = this.persistenceRequired || this.gale$eventDriven_requiresCustomPersistence;
+        if (this.gale$eventDriven_isPersistent != newValue) {
+            this.gale$eventDriven_isPersistent = newValue;
+            this.gale$eventDriven_checkDespawnPlayerTrackingMode_update(); // Gale - Event-driven - Mob.checkDespawnPlayerTrackingMode
+        }
+    }
+    // Gale end - Event-driven - Mob.isPersistent
+
+    // Gale start - Optimize Mob.checkDespawn()
+    protected final int gale$precompute_typeCategoryOrdinal = this.getType().getCategory().ordinal(); // Gale - Pre-compute - Mob.getType().getCategory.ordinal()
+
+    /**
+     * The index in {@link Level#gale$checkDespawn_spatialGridIndices}
+     * that should be used by this mob,
+     * for the soft range.
+     */
+    protected final int gale$checkDespawn_softLevelSpatialGridIndicesIndex = this.gale$precompute_typeCategoryOrdinal << 1; // Gale - Pre-compute - Mob.getType().getCategory.ordinal()
+
+    /**
+     * Analogous to {@link #gale$checkDespawn_softLevelSpatialGridIndicesIndex},
+     * but for the hard range.
+     */
+    protected final int gale$checkDespawn_hardLevelSpatialGridIndicesIndex = this.gale$checkDespawn_softLevelSpatialGridIndicesIndex + 1;
+
+    /**
+     * Whether this mob needs to track players for {@link #checkDespawn}:
+     * <ul>
+     *     <li>0 = none</li>
+     *     <li>1 = only for the {@link io.papermc.paper.configuration.WorldConfiguration.Entities.Spawning.DespawnRangePair#soft()} distance</li>
+     *     <li>2 = for both {@link io.papermc.paper.configuration.WorldConfiguration.Entities.Spawning.DespawnRangePair#soft()} and
+     *     {@link io.papermc.paper.configuration.WorldConfiguration.Entities.Spawning.DespawnRangePair#hard()} distances</li>
+     * </ul>
+     */
+    public int gale$eventDriven_checkDespawnPlayerTrackingMode;
+
+    public void gale$eventDriven_checkDespawnPlayerTrackingMode_update() {
+        int newValue = this.gale$eventDriven_checkDespawnPlayerTrackingMode_compute();
+        if (this.gale$eventDriven_checkDespawnPlayerTrackingMode != newValue) {
+            this.gale$eventDriven_checkDespawnPlayerTrackingMode = newValue;
+            if (newValue == 0) {
+                this.gale$checkDespawn_lastSoftPlayerSlot = -1;
+                this.gale$checkDespawn_lastSoftPlayerAffectsSpawningId = -1;
+            }
+        }
+    }
+
+    private byte gale$eventDriven_checkDespawnPlayerTrackingMode_compute() {
+        if (/*!this.gale$eventDriven_isInTickList || */this.gale$eventDriven_mustDespawnBecauseOfPeacefulDifficulty || this.gale$eventDriven_isPersistent || this.level().gale$eventDriven_affectsSpawningSelectorPlayerCount == 0) {
+            return 0;
+        } else if (!this.gale$canRemoveWhenFarAway()) {
+            return 1;
+        } else {
+            return 2;
+        }
+    }
+
+    // @Override
+    // public void gale$eventDriven_isInTickList_set(boolean newValue) {
+    //     boolean oldValue = this.gale$eventDriven_isInTickList;
+    //     if (oldValue != newValue) {
+    //         super.gale$eventDriven_isInTickList_set(newValue);
+    //         this.gale$eventDriven_checkDespawnPlayerTrackingMode_update();
+    //     }
+    // }
 
     @Override
     public void checkDespawn() {
-        if (this.level().getDifficulty() == Difficulty.PEACEFUL && !this.getType().isAllowedInPeaceful()) {
+        if (this.gale$eventDriven_mustDespawnBecauseOfPeacefulDifficulty) { // Gale - Event-driven - Mob.mustDespawnBecauseOfPeacefulDifficulty
             this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
-        } else if (!this.isPersistenceRequired() && !this.requiresCustomPersistence()) {
-            Entity player = this.level().findNearbyPlayer(this, -1.0, EntitySelector.PLAYER_AFFECTS_SPAWNING); // Paper - Affects Spawning API
+        } else if (!this.gale$eventDriven_isPersistent) { // Gale - Event-driven - Mob.isPersistent
+            if (!org.galemc.gale.async.SimulationFlag.REAL) return; // Gale - Speculative execution based on next tick simulation - Simulate next tick
+            if (this.gale$eventDriven_checkDespawnPlayerTrackingMode != 0) {
+                Level level = this.level();
+                if (level.gale$checkDespawn_usingSpatialGrid) {
+                    if (!this.gale$checkDespawn_innerWithLastSoftPlayerBySpatialGridSlot()) {
+                        this.gale$checkDespawn_innerWithSpatialGrids();
+                    }
+                } else {
+                    if (level.gale$eventDriven_affectsSpawningSelectorPlayerCount > 1) {
+                        if (!this.gale$checkDespawn_innerWithLastSoftPlayerByAffectsSpawningId()) {
+                            this.gale$checkDespawn_innerWithPlayerList();
+                        }
+                    } else {
+                        this.gale$checkDespawn_innerWithPlayerList();
+                    }
+                }
+            }
+        } else {
+            if (!org.galemc.gale.async.SimulationFlag.REAL) return; // Gale - Speculative execution based on next tick simulation - Simulate next tick
+            this.noActionTime = 0;
+        }
+    }
+
+    protected int gale$checkDespawn_lastSoftPlayerSlot = -1;
+    protected int gale$checkDespawn_lastSoftPlayerAffectsSpawningId = -1;
+
+    private boolean gale$checkDespawn_innerWithLastSoftPlayerBySpatialGridSlot() {
+        if (this.gale$checkDespawn_lastSoftPlayerSlot == -1) return false;
+        Level level = this.level();
+        org.galemc.gale.util.collection.spatialgrid.despawn.point.SpatialGrid softGrid = level.gale$checkDespawn_spatialGrids[level.gale$checkDespawn_spatialGridIndices[this.gale$checkDespawn_softLevelSpatialGridIndicesIndex]];
+        if (softGrid.containsKey(this.gale$checkDespawn_lastSoftPlayerSlot)) {
+            double playerX = softGrid.getX(this.gale$checkDespawn_lastSoftPlayerSlot);
+            double playerY = softGrid.getY(this.gale$checkDespawn_lastSoftPlayerSlot);
+            double playerZ = softGrid.getZ(this.gale$checkDespawn_lastSoftPlayerSlot);
+            Vec3 position = this.position;
+            double x = position.x;
+            double y = position.y;
+            double z = position.z;
+            double dx = playerX - x;
+            double dy = Math.abs(playerY - y);
+            double dz = playerZ - z;
+            double dxSqr = dx * dx;
+            double dySqr = dy * dy;
+            double dzSqr = dz * dz;
+            final io.papermc.paper.configuration.WorldConfiguration.Entities.Spawning.DespawnRangePair despawnRangePair = level.gale$checkDespawn_despawnRangePairs[this.gale$precompute_typeCategoryOrdinal]; // Gale - Pre-compute - Mob.getType().getCategory.ordinal()
+            io.papermc.paper.configuration.type.DespawnRange despawnRange = despawnRangePair.soft();
+            final io.papermc.paper.configuration.type.DespawnRange.Shape shape = level.gale$checkDespawn_despawnRangeShape; // Gale - Pre-compute - Mob.checkDespawn() despawn range shape
+            if (!despawnRange.shouldDespawn(shape, dxSqr, dySqr, dzSqr, dy)) {
+                this.noActionTime = 0;
+                return true;
+            }
+        }
+        this.gale$checkDespawn_lastSoftPlayerSlot = -1;
+        return false;
+    }
+
+    private boolean gale$checkDespawn_innerWithLastSoftPlayerByAffectsSpawningId() {
+        if (this.gale$checkDespawn_lastSoftPlayerAffectsSpawningId == -1) return false;
+        Level level = this.level();
+        if (level.gale$eventDriven_affectsSpawningSelectorPlayersById.length > this.gale$checkDespawn_lastSoftPlayerAffectsSpawningId) {
+            Player player = level.gale$eventDriven_affectsSpawningSelectorPlayersById[this.gale$checkDespawn_lastSoftPlayerAffectsSpawningId];
             if (player != null) {
-                // Paper start - Configurable despawn distances
-                final io.papermc.paper.configuration.WorldConfiguration.Entities.Spawning.DespawnRangePair despawnRangePair = this.level().paperConfig().entities.spawning.despawnRanges.get(this.getType().getCategory());
-                final io.papermc.paper.configuration.type.DespawnRange.Shape shape = this.level().paperConfig().entities.spawning.despawnRangeShape;
-                final double dy = Math.abs(player.getY() - this.getY());
-                final double dySqr = Mth.square(dy);
-                final double dxSqr = Mth.square(player.getX() - this.getX());
-                final double dzSqr = Mth.square(player.getZ() - this.getZ());
-                final double distanceSquared = dxSqr + dzSqr + dySqr;
-                // Despawn if hard/soft limit is exceeded
-                if (despawnRangePair.hard().shouldDespawn(shape, dxSqr, dySqr, dzSqr, dy) && this.removeWhenFarAway(distanceSquared)) {
-                    this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
-                }
-
-                if (despawnRangePair.soft().shouldDespawn(shape, dxSqr, dySqr, dzSqr, dy)) {
-                    if (this.noActionTime > 600 && this.random.nextInt(800) == 0 && this.removeWhenFarAway(distanceSquared)) {
-                        this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
-                    }
-                } else {
-                    // Paper end - Configurable despawn distances
+                double playerX = player.getX();
+                double playerY = player.getY();
+                double playerZ = player.getZ();
+                Vec3 position = this.position;
+                double x = position.x;
+                double y = position.y;
+                double z = position.z;
+                double dx = playerX - x;
+                double dy = Math.abs(playerY - y);
+                double dz = playerZ - z;
+                double dxSqr = dx * dx;
+                double dySqr = dy * dy;
+                double dzSqr = dz * dz;
+                final io.papermc.paper.configuration.WorldConfiguration.Entities.Spawning.DespawnRangePair despawnRangePair = level.gale$checkDespawn_despawnRangePairs[this.gale$precompute_typeCategoryOrdinal]; // Gale - Pre-compute - Mob.getType().getCategory.ordinal()
+                io.papermc.paper.configuration.type.DespawnRange despawnRange = despawnRangePair.soft();
+                final io.papermc.paper.configuration.type.DespawnRange.Shape shape = level.gale$checkDespawn_despawnRangeShape; // Gale - Pre-compute - Mob.checkDespawn() despawn range shape
+                if (!despawnRange.shouldDespawn(shape, dxSqr, dySqr, dzSqr, dy)) {
                     this.noActionTime = 0;
-                }
-            }
+                    return true;
+                }
+            }
+        }
+        this.gale$checkDespawn_lastSoftPlayerAffectsSpawningId = -1;
+        return false;
+    }
+
+    protected void gale$checkDespawn_innerWithPlayerList() { // More like the original implementation
+        Level level = this.level();
+        Vec3 position = this.position;
+        double x = position.x;
+        double y = position.y;
+        double z = position.z;
+        Player[] affectsSpawningPlayers = level.gale$eventDriven_affectsSpawningSelectorPlayersDense;
+        int affectsSpawningPlayerCount = level.gale$eventDriven_affectsSpawningSelectorPlayerCount;
+        net.minecraft.world.entity.player.Player player = null;
+        double dxSqr = 0;
+        double dyUnsqr = 0;
+        double dySqr = 0;
+        double dzSqr = 0;
+        double distanceSquared = Double.MAX_VALUE;
+        for (int i = 0; i < affectsSpawningPlayerCount; i++) {
+            Player affectsSpawningPlayer = affectsSpawningPlayers[i];
+            double pdxSqr = Mth.square(affectsSpawningPlayer.getX() - x);
+            double pdyUnsqr = affectsSpawningPlayer.getY() - y;
+            double pdySqr = Mth.square(pdyUnsqr);
+            double pdzSqr = Mth.square(affectsSpawningPlayer.getZ() - z);
+            double pdistanceSquared = pdxSqr + pdzSqr + pdySqr;
+            if (pdistanceSquared < distanceSquared) {
+                player = affectsSpawningPlayer;
+                dxSqr = pdxSqr;
+                dyUnsqr = pdyUnsqr;
+                dySqr = pdySqr;
+                dzSqr = pdzSqr;
+                distanceSquared = pdistanceSquared;
+            }
+        }
+        double dy = Math.abs(dyUnsqr);
+        // Paper start - Configurable despawn distances
+        final io.papermc.paper.configuration.WorldConfiguration.Entities.Spawning.DespawnRangePair despawnRangePair = level.gale$checkDespawn_despawnRangePairs[this.gale$precompute_typeCategoryOrdinal]; // Gale - Pre-compute - Mob.getType().getCategory.ordinal()
+        final io.papermc.paper.configuration.type.DespawnRange.Shape shape = level.gale$checkDespawn_despawnRangeShape; // Gale - Pre-compute - Mob.checkDespawn() despawn range shape
+        // Despawn if hard/soft limit is exceeded
+        if (this.gale$canRemoveWhenFarAway() && despawnRangePair.hard().shouldDespawn(shape, dxSqr, dySqr, dzSqr, dy)) {
+            this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
+        }
+
+        if (this.noActionTime != 0) {
+            if (despawnRangePair.soft().shouldDespawn(shape, dxSqr, dySqr, dzSqr, dy)) {
+                if (this.gale$canRemoveWhenFarAway() && this.noActionTime > 600 && this.random.nextInt(800) == 0) {
+                    this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
+                }
+            } else {
+                this.gale$checkDespawn_lastSoftPlayerAffectsSpawningId = player.gale$eventDriven_affectsSpawningSelectorLevelId;
+                // Paper end - Configurable despawn distances
+                this.noActionTime = 0;
+            }
+        }
+    }
+
+    protected void gale$checkDespawn_innerWithSpatialGrids() {
+        // Paper start - Configurable despawn distances
+        // Despawn if hard/soft limit is exceeded
+
+        Level level = this.level();
+        org.galemc.gale.util.collection.spatialgrid.despawn.point.SpatialGrid[] grids = level.gale$checkDespawn_spatialGrids;
+        int[] gridIndices = level.gale$checkDespawn_spatialGridIndices;
+        Vec3 position = this.position;
+        double x = position.x;
+        double y = position.y;
+        double z = position.z;
+        BlockPos blockPosition = this.blockPosition;
+        int blockX = blockPosition.getX();
+        int blockY = blockPosition.getY();
+        int blockZ = blockPosition.getZ();
+
+        if (this.noActionTime == 0) {
+
+            // Check in hard range if necessary
+            if (this.gale$eventDriven_checkDespawnPlayerTrackingMode == 2) {
+                int hardGridIndex = gridIndices[this.gale$checkDespawn_hardLevelSpatialGridIndicesIndex];
+                org.galemc.gale.util.collection.spatialgrid.despawn.point.SpatialGrid hardGrid = grids[hardGridIndex];
+                if (hardGrid.anyIn(x, y, z, blockX, blockY, blockZ) == -1) {
+                    this.gale$checkDespawn_lastSoftPlayerSlot = -1;
+                    this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
+                }
+            }
+
         } else {
-            this.noActionTime = 0;
+
+            // Check in soft range
+            int softGridIndex = gridIndices[this.gale$checkDespawn_softLevelSpatialGridIndicesIndex];
+            org.galemc.gale.util.collection.spatialgrid.despawn.point.SpatialGrid softGrid = grids[softGridIndex];
+            int softSlot = softGrid.anyIn(x, y, z, blockX, blockY, blockZ);
+            this.gale$checkDespawn_lastSoftPlayerSlot = softSlot;
+
+            if (softSlot == -1) {
+
+                // Check in hard range if necessary
+                if (this.gale$eventDriven_checkDespawnPlayerTrackingMode == 2) {
+                    int hardGridIndex = gridIndices[this.gale$checkDespawn_hardLevelSpatialGridIndicesIndex];
+                    org.galemc.gale.util.collection.spatialgrid.despawn.point.SpatialGrid hardGrid = grids[hardGridIndex];
+                    if (hardGrid.anyIn(x, y, z, blockX, blockY, blockZ) == -1) {
+                        this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
+                    }
+                }
+
+                if (this.gale$canRemoveWhenFarAway() && this.noActionTime > 600 && this.random.nextInt(800) == 0) {
+                    this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
+                }
+
+            } else {
+
+                // Paper end - Configurable despawn distances
+                this.noActionTime = 0;
+
+            }
+
         }
+
     }
+    // Gale end - Optimize Mob.checkDespawn()
 
     @Override
     protected final void serverAiStep() {
@@ -1275,6 +_,7 @@
 
     public void setPersistenceRequired() {
         this.persistenceRequired = true;
+        this.gale$eventDriven_isPersistent_update(); // Gale - Event-driven - Mob.isPersistent
     }
 
     @Override
@@ -1487,7 +_,21 @@
 
     @Override
     public void setLeashData(final Leashable.@Nullable LeashData leashData) {
-        this.leashData = leashData;
+        // Gale start - Event-driven - Mob.isLeashed
+        if (this.leashData != leashData) {
+            if (this.leashData != null && this.leashDataListenerReference != null) {
+                this.leashData.removeLeashHolderListener(this.leashDataListenerReference);
+            }
+            this.leashData = leashData;
+            this.gale$eventDriven_isLeashed_update();
+            if (leashData != null) {
+                if (this.leashDataListenerReference == null) {
+                    this.leashDataListenerReference = new java.lang.ref.WeakReference<>(this);
+                }
+                leashData.addLeashHolderListener(this.leashDataListenerReference);
+            }
+        }
+        // Gale end - Event-driven - Mob.isLeashed
     }
 
     @Override
