--- a/net/minecraft/world/level/block/BaseRailBlock.java
+++ b/net/minecraft/world/level/block/BaseRailBlock.java
@@ -35,7 +_,7 @@
     }
 
     public static boolean isRail(final BlockState state) {
-        return state.is(BlockTags.RAILS) && state.getBlock() instanceof BaseRailBlock;
+        return state.isAny(org.dreeam.leaf.util.BlockMasks.RAILS_TAG_AND_BASE_RAIL_CLASS); // Gale - Pre-compute - Common BlockState predicates - Apply
     }
 
     protected BaseRailBlock(final boolean isStraight, final BlockBehaviour.Properties properties) {
