--- a/net/minecraft/world/level/block/Block.java
+++ b/net/minecraft/world/level/block/Block.java
@@ -79,7 +_,7 @@
     public static final MapCodec<Block> CODEC = simpleCodec(Block::new);
     private static final Logger LOGGER = LogUtils.getLogger();
     private final Holder.Reference<Block> builtInRegistryHolder = BuiltInRegistries.BLOCK.createIntrusiveHolder(this);
-    public static final IdMapper<BlockState> BLOCK_STATE_REGISTRY = new IdMapper<>();
+    public static final org.galemc.gale.registry.blockstate.BlockStateIdMapper BLOCK_STATE_REGISTRY = new org.galemc.gale.registry.blockstate.BlockStateIdMapper(); // Gale - Optimize global block state palette
     private static final LoadingCache<VoxelShape, Boolean> SHAPE_FULL_BLOCK_CACHE = CacheBuilder.newBuilder()
         .maximumSize(512L)
         .weakKeys()
@@ -148,8 +_,7 @@
             return 0;
         }
 
-        int id = BLOCK_STATE_REGISTRY.getId(blockState);
-        return id == -1 ? 0 : id;
+        return blockState.indexInRegistry; // Gale - Optimize global block state palette
     }
 
     public static BlockState stateById(final int idWithData) {
