--- a/net/minecraft/world/level/block/FireBlock.java
+++ b/net/minecraft/world/level/block/FireBlock.java
@@ -136,7 +_,7 @@
         if (!this.canBurn(belowState) && !belowState.isFaceSturdy(level, below, Direction.UP)) {
             BlockState result = this.defaultBlockState();
 
-            for (Direction direction : Direction.values()) {
+            for (Direction direction : Direction.gale$precompute_VALUES) { // Gale - Pre-compute - Direction.values()
                 BooleanProperty property = PROPERTY_BY_DIRECTION.get(direction);
                 if (property != null) {
                     result = result.setValue(property, this.canBurn(level.getBlockState(pos.relative(direction))));
@@ -306,7 +_,7 @@
     }
 
     private boolean isValidFireLocation(final BlockGetter level, final BlockPos pos) {
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.gale$precompute_VALUES) { // Gale - Pre-compute - Direction.values()
             if (this.canBurn(level.getBlockState(pos.relative(direction)))) {
                 return true;
             }
@@ -322,7 +_,7 @@
 
         int odds = 0;
 
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.gale$precompute_VALUES) { // Gale - Pre-compute - Direction.values()
             BlockState blockState = level.getBlockState(pos.relative(direction));
             odds = Math.max(this.getIgniteOdds(blockState), odds);
         }
