--- a/net/minecraft/world/level/block/RespawnAnchorBlock.java
+++ b/net/minecraft/world/level/block/RespawnAnchorBlock.java
@@ -159,24 +_,31 @@
         FluidState fluidBelow = level.getFluidState(pos.below());
         return !fluidBelow.is(FluidTags.WATER);
     }
 
     private void explode(final BlockState state, final ServerLevel level, final BlockPos pos) {
         org.bukkit.block.BlockState blockState = org.bukkit.craftbukkit.block.CraftBlock.at(level, pos).getState(); // CraftBukkit - capture BlockState before remove block
         level.removeBlock(pos, false);
-        boolean anyWaterNeighbors = Direction.Plane.HORIZONTAL.stream().map(pos::relative).anyMatch(neighborPos -> isWaterThatWouldFlow(neighborPos, level));
-        final boolean inWater = anyWaterNeighbors || level.getFluidState(pos.above()).is(FluidTags.WATER);
         ExplosionDamageCalculator damageCalculator = new ExplosionDamageCalculator() {
             @Override
             public Optional<Float> getBlockExplosionResistance(
                 final Explosion explosion, final BlockGetter levelx, final BlockPos testPos, final BlockState block, final FluidState fluid
             ) {
-                return testPos.equals(pos) && inWater
+                return testPos.equals(pos) && hasWaterFlow(level, pos)
                     ? Optional.of(Blocks.WATER.getExplosionResistance())
                     : super.getBlockExplosionResistance(explosion, levelx, testPos, block, fluid);
             }
         };
         Vec3 boomPos = Vec3.atCenterOf(pos);
         level.explode(null, level.damageSources().badRespawnPointExplosion(boomPos).causingBlockSnapshot(blockState), damageCalculator, boomPos, 5.0F, true, Level.ExplosionInteraction.BLOCK); // CraftBukkit - add state
     }
 
+    // Leaf start - Optimize respawn anchor explosion
+    private static boolean hasWaterFlow(final Level level, final BlockPos secondPos) {
+        for (Direction direction : Direction.Plane.HORIZONTAL) {
+            final BlockPos neighborPos = secondPos.relative(direction);
+            if (isWaterThatWouldFlow(neighborPos, level)) {
+                return true;
+            }
+        }
+        return level.getFluidState(secondPos.above()).is(FluidTags.WATER);
+    }
+    // Leaf end - Optimize respawn anchor explosion
+
     public static boolean canSetSpawn(final ServerLevel level, final BlockPos pos) {
