--- a/net/minecraft/world/phys/shapes/CubePointRange.java
+++ b/net/minecraft/world/phys/shapes/CubePointRange.java
@@ -4,6 +_,7 @@
 
 public class CubePointRange extends AbstractDoubleList {
     private final int parts;
+    private final double scale; // Gale - Replace division by multiplication - Pre-compute scale
 
     public CubePointRange(final int parts) {
         if (parts <= 0) {
@@ -11,11 +_,12 @@
         }
 
         this.parts = parts;
+        this.scale = 1.0D / parts; // Gale - Replace division by multiplication - Pre-compute scale
     }
 
     @Override
     public double getDouble(final int index) {
-        return (double)index / this.parts;
+        return index * this.scale; // Gale - Replace division by multiplication - Pre-compute scale
     }
 
     @Override
