--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -236,7 +_,12 @@
         // Spigot end
         // Paper start - initialize global and world-defaults configuration
         this.paperConfigurations.initializeGlobalConfiguration(this.registryAccess());
+        net.minecraft.world.level.block.state.BlockBehaviour.gale$eventDriven_updateConfigBased(); // Gale - Event-driven - BlockBehaviour.getPistonPushReaction(), BlockBehaviour.isDestroyable() - Initialize config-dependent block-state caches
         this.paperConfigurations.initializeWorldDefaultsConfiguration(this.registryAccess());
+        // Gale start - Configuration - Initialize configuration
+        // this.galeConfigurations.initializeGlobalConfiguration(this.registryAccess()); // TODO Gale configuration temp disabled
+        // this.galeConfigurations.initializeWorldDefaultsConfiguration(this.registryAccess()); // TODO Gale configuration temp disabled
+        // Gale end - Configuration - Initialize configuration
         // Paper end - initialize global and world-defaults configuration
         this.server.spark.enableEarlyIfRequested(); // Paper - spark
         // Paper start - fix converting txt to json file; convert old users earlier after PlayerList creation but before file load/save
@@ -402,7 +_,7 @@
     @Override
     protected void tickServer(final BooleanSupplier haveTime) {
         super.tickServer(haveTime);
-        if (this.jsonRpcServer != null) {
+        if (org.galemc.gale.async.SimulationFlag.REAL) if (this.jsonRpcServer != null) { // Gale - Speculative execution based on next tick simulation - Simulate next tick
             this.jsonRpcServer.tick();
         }
     }
@@ -917,7 +_,11 @@
 
     @Override
     public int getScaledTrackingDistance(final int range) {
-        return this.entityBroadcastRangePercentage() * range / 100;
+        // Gale start - Local code optimization - DedicatedServer.getScaledTrackingDistance()
+        int p = this.getProperties().entityBroadcastRangePercentage.get();
+        if (p == 100) return range;
+        return p * range / 100;
+        // Gale end - Local code optimization - DedicatedServer.getScaledTrackingDistance()
     }
 
     public int entityBroadcastRangePercentage() {
