--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -237,6 +_,10 @@
         // Paper start - initialize global and world-defaults configuration
         this.paperConfigurations.initializeGlobalConfiguration(this.registryAccess());
         this.paperConfigurations.initializeWorldDefaultsConfiguration(this.registryAccess());
+        // Gale start - initialize gale configuration
+        this.galeConfigurations.initializeGlobalConfiguration(this.registryAccess());
+        this.galeConfigurations.initializeWorldDefaultsConfiguration(this.registryAccess());
+        // Gale end - initialize gale configuration
         // Paper end - initialize global and world-defaults configuration
         this.server.spark.enableEarlyIfRequested(); // Paper - spark
         // Paper start - fix converting txt to json file; convert old users earlier after PlayerList creation but before file load/save