--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -918,6 +_,7 @@
 
             chunk.setNoiseBiome(QuartPos.fromBlock(x), QuartPos.fromBlock(y), QuartPos.fromBlock(z), biome);
             chunk.markUnsaved(); // SPIGOT-2890
+            this.world.getBiomeManager().invalidateBiomeCache(); // Gale - Cache - BiomeManager.getBiome() - Invalidate cache on writes
         }
     }
 
