--- a/net/minecraft/world/TickRateManager.java
+++ b/net/minecraft/world/TickRateManager.java
@@ -61,6 +_,6 @@
     }
 
     public boolean isEntityFrozen(final Entity entity) {
-        return !this.runsNormally() && !(entity instanceof Player) && entity.countPlayerPassengers() <= 0;
+        return !this.runsNormally() && !(entity instanceof Player) && entity.hasZeroPlayerPassengers(); // Gale - Optimize player passenger counting
     }
 }
