--- a/net/minecraft/world/level/redstone/DefaultRedstoneWireEvaluator.java
+++ b/net/minecraft/world/level/redstone/DefaultRedstoneWireEvaluator.java
@@ -1,7 +_,5 @@
 package net.minecraft.world.level.redstone;
 
-import com.google.common.collect.Sets;
-import java.util.Set;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.world.level.Level;
@@ -25,12 +_,9 @@
                 level.setBlock(pos, state.setValue(RedStoneWireBlock.POWER, targetStrength), Block.UPDATE_CLIENTS);
             }
 
-            Set<BlockPos> toUpdate = Sets.newHashSet();
-            toUpdate.add(pos);
-
-            for (Direction direction : Direction.values()) {
-                toUpdate.add(pos.relative(direction));
-            }
+            // Gale start - optimize redstone wire update - replace HashSet with fixed array
+            BlockPos[] toUpdate = {pos, pos.relative(Direction.DOWN), pos.relative(Direction.UP), pos.relative(Direction.NORTH), pos.relative(Direction.SOUTH), pos.relative(Direction.WEST), pos.relative(Direction.EAST)};
+            // Gale end - optimize redstone wire update - replace HashSet with fixed array
 
             for (BlockPos blockPos : toUpdate) {
                 level.updateNeighborsAt(blockPos, this.wireBlock);
