--- a/net/minecraft/world/level/redstone/DefaultRedstoneWireEvaluator.java
+++ b/net/minecraft/world/level/redstone/DefaultRedstoneWireEvaluator.java
@@ -1,7 +_,5 @@
 package net.minecraft.world.level.redstone;
 
-import com.google.common.collect.Sets;
-import java.util.Set;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.world.level.Level;
@@ -25,12 +_,16 @@
                 level.setBlock(pos, state.setValue(RedStoneWireBlock.POWER, targetStrength), Block.UPDATE_CLIENTS);
             }
 
-            Set<BlockPos> toUpdate = Sets.newHashSet();
-            toUpdate.add(pos);
-
-            for (Direction direction : Direction.values()) {
-                toUpdate.add(pos.relative(direction));
-            }
+            // Gale start - optimize redstone wire update - replace HashSet with fixed array
+            BlockPos[] toUpdate = new BlockPos[7];
+            toUpdate[0] = pos;
+            toUpdate[1] = pos.relative(Direction.NORTH);
+            toUpdate[2] = pos.relative(Direction.EAST);
+            toUpdate[3] = pos.relative(Direction.SOUTH);
+            toUpdate[4] = pos.relative(Direction.WEST);
+            toUpdate[5] = pos.relative(Direction.UP);
+            toUpdate[6] = pos.relative(Direction.DOWN);
+            // Gale end - optimize redstone wire update - replace HashSet with fixed array
 
             for (BlockPos blockPos : toUpdate) {
                 level.updateNeighborsAt(blockPos, this.wireBlock);
