--- a/net/minecraft/world/level/redstone/RedstoneWireEvaluator.java
+++ b/net/minecraft/world/level/redstone/RedstoneWireEvaluator.java
@@ -28,13 +_,16 @@
 
     protected int getIncomingWireSignal(final Level level, final BlockPos pos) {
         int wireSignal = 0;
+        // Gale start - hoist pos.above() out of loop
+        final BlockPos abovePos = pos.above();
+        final boolean aboveIsConductor = level.getBlockState(abovePos).isRedstoneConductor(level, abovePos);
+        // Gale end - hoist pos.above() out of loop
 
         for (Direction direction : Direction.Plane.HORIZONTAL) {
             BlockPos neighborPos = pos.relative(direction);
             BlockState neighborState = level.getBlockState(neighborPos);
             wireSignal = Math.max(wireSignal, this.getWireSignal(neighborPos, neighborState));
-            BlockPos abovePos = pos.above();
-            if (neighborState.isRedstoneConductor(level, neighborPos) && !level.getBlockState(abovePos).isRedstoneConductor(level, abovePos)) {
+            if (neighborState.isRedstoneConductor(level, neighborPos) && !aboveIsConductor) {
                 BlockPos aboveNeighborPos = neighborPos.above();
                 wireSignal = Math.max(wireSignal, this.getWireSignal(aboveNeighborPos, level.getBlockState(aboveNeighborPos)));
             } else if (!neighborState.isRedstoneConductor(level, neighborPos)) {
@@ -45,4 +_,4 @@
 
         return Math.max(0, wireSignal - 1);
     }
 }
