--- a/net/minecraft/world/level/biome/BiomeManager.java
+++ b/net/minecraft/world/level/biome/BiomeManager.java
@@ -14,8 +_,30 @@
     private static final int ZOOM_MASK = 3;
     private final BiomeManager.NoiseBiomeSource noiseBiomeSource;
     private final long biomeZoomSeed;
+    // Gale start - Cache - BiomeManager.getBiome()
+    private final long[] biomeCacheKeys;
+    private final int[] biomeCacheVersions;
+    private final Holder<Biome>[] biomeCacheValues;
+    private int cacheVersion;
+    // Gale end - Cache - BiomeManager.getBiome()
 
     public BiomeManager(final BiomeManager.NoiseBiomeSource noiseBiomeSource, final long seed) {
+        // Gale start - Cache - BiomeManager.getBiome()
+        this(noiseBiomeSource, seed, false);
+    }
+
+    public BiomeManager(final BiomeManager.NoiseBiomeSource noiseBiomeSource, final long seed, boolean initializeCache) {
+        if (initializeCache) {
+            this.biomeCacheKeys = new long[1 << 15];
+            java.util.Arrays.fill(this.biomeCacheKeys, Long.MAX_VALUE); // Long.MAX_VALUE is not a possible BlockPos packed long
+            this.biomeCacheVersions = new int[1 << 15];
+            this.biomeCacheValues = new Holder[1 << 15];
+        } else {
+            this.biomeCacheKeys = null;
+            this.biomeCacheVersions = null;
+            this.biomeCacheValues = null;
+        }
+        // Gale end - Cache - BiomeManager.getBiome()
         this.noiseBiomeSource = noiseBiomeSource;
         this.biomeZoomSeed = seed;
     }
@@ -28,6 +_,28 @@
         return new BiomeManager(biomeSource, this.biomeZoomSeed);
     }
 
+    // Gale start - Cache - BiomeManager.getBiome()
+    public Holder<Biome> getBiomeCached(final BlockPos pos) {
+        // Look up in cache
+        long cacheKey = pos.asLong();
+        int cacheIndex = ((int) it.unimi.dsi.fastutil.HashCommon.mix(cacheKey)) & ((1 << 15) - 1);
+        if (this.biomeCacheKeys[cacheIndex] == cacheKey && this.biomeCacheVersions[cacheIndex] == this.cacheVersion) {
+            return this.biomeCacheValues[cacheIndex];
+        }
+
+        // Compute and store in cache
+        Holder<Biome> computedBiome = this.getBiome(pos);
+        this.biomeCacheKeys[cacheIndex] = cacheKey;
+        this.biomeCacheVersions[cacheIndex] = this.cacheVersion;
+        this.biomeCacheValues[cacheIndex] = computedBiome;
+        return computedBiome;
+    }
+
+    public void invalidateBiomeCache() {
+        this.cacheVersion++;
+    }
+    // Gale end - Cache - BiomeManager.getBiome()
+
     public Holder<Biome> getBiome(final BlockPos pos) {
         int absX = pos.getX() - 2;
         int absY = pos.getY() - 2;
@@ -101,7 +_,20 @@
     }
 
     private static double getFiddle(final long rval) {
-        return (double)(((rval >> 24) & (1024 - 1)) - (1024/2)) * (0.9 / 1024.0); // Paper - avoid floorMod, fp division, and fp subtraction
+        // Gale start - Pre-compute - BiomeManager.getFiddle()
+        return PRECOMPUTED_FIDDLE_TABLE[(int) (rval >>> 24) & 1023];
+    }
+
+    private static final double[] PRECOMPUTED_FIDDLE_TABLE = new double[1024];
+    static {
+        for (int i = 0; i < 1024; i++) {
+            PRECOMPUTED_FIDDLE_TABLE[i] = precomputeFiddle(i);
+        }
+    }
+
+    private static double precomputeFiddle(final int i) {
+        // Gale end - Pre-compute - BiomeManager.getFiddle()
+        return (double)(i - (1024/2)) * (0.9 / 1024.0); // Paper - avoid floorMod, fp division, and fp subtraction
     }
 
     public interface NoiseBiomeSource {
