--- a/net/minecraft/world/level/biome/BiomeManager.java
+++ b/net/minecraft/world/level/biome/BiomeManager.java
@@ -101,7 +_,20 @@
     }
 
     private static double getFiddle(final long rval) {
-        return (double)(((rval >> 24) & (1024 - 1)) - (1024/2)) * (0.9 / 1024.0); // Paper - avoid floorMod, fp division, and fp subtraction
+        // Gale start - Pre-compute - BiomeManager.getFiddle()
+        return PRECOMPUTED_FIDDLE_TABLE[(int) (rval >>> 24) & 1023];
+    }
+
+    private static final double[] PRECOMPUTED_FIDDLE_TABLE = new double[1024];
+    static {
+        for (int i = 0; i < 1024; i++) {
+            PRECOMPUTED_FIDDLE_TABLE[i] = precomputeFiddle(i);
+        }
+    }
+
+    private static double precomputeFiddle(final int i) {
+        // Gale end - Pre-compute - BiomeManager.getFiddle()
+        return (double)(i - (1024/2)) * (0.9 / 1024.0); // Paper - avoid floorMod, fp division, and fp subtraction
     }
 
     public interface NoiseBiomeSource {
