--- a/net/minecraft/world/level/pathfinder/NodeEvaluator.java
+++ b/net/minecraft/world/level/pathfinder/NodeEvaluator.java
@@ -96,6 +_,12 @@
     }
 
     public static boolean isBurningBlock(final BlockState blockState) {
+        // Gale start - Pre-compute - Common BlockState predicates - Apply
+        return blockState.isAny(org.dreeam.leaf.util.BlockMasks.IS_BURNING_BLOCK);
+    }
+
+    public static boolean gale$precompute_isBurningBlock_compute(final BlockState blockState) {
+        // Gale end - Pre-compute - Common BlockState predicates - Apply
         return blockState.is(BlockTags.FIRE)
             || blockState.is(Blocks.LAVA)
             || blockState.is(Blocks.MAGMA_BLOCK)
