--- a/net/minecraft/resources/Identifier.java
+++ b/net/minecraft/resources/Identifier.java
@@ -26,6 +_,10 @@
     public static final String PAPER_NAMESPACE = "paper"; // Paper
     private final String namespace;
     private final String path;
+    // Gale start - Cache - Identifier.hashCode() and Identifier.toString()
+    private int gale$cached_hashCode;
+    private @Nullable String gale$cached_string;
+    // Gale end - Cache - Identifier.hashCode() and Identifier.toString()
 
     private Identifier(final String namespace, final String path) {
         assert isValidNamespace(namespace);
@@ -131,7 +_,14 @@
 
     @Override
     public String toString() {
-        return this.namespace + ":" + this.path;
+        // Gale start - Cache - Identifier.hashCode() and Identifier.toString()
+        String string = this.gale$cached_string;
+        if (string == null) {
+            string = this.namespace + ":" + this.path;
+            this.gale$cached_string = string;
+        }
+        return string;
+        // Gale end - Cache - Identifier.hashCode() and Identifier.toString()
     }
 
     @Override
@@ -141,7 +_,15 @@
 
     @Override
     public int hashCode() {
-        return 31 * this.namespace.hashCode() + this.path.hashCode();
+        // Gale start - Cache - Identifier.hashCode() and Identifier.toString()
+        int hashCode = this.gale$cached_hashCode;
+        if (hashCode == 0) {
+            hashCode = 1 + this.namespace.hashCode();
+            hashCode = 31 * hashCode + this.path.hashCode();
+            this.gale$cached_hashCode = hashCode;
+        }
+        return hashCode;
+        // Gale end - Cache - Identifier.hashCode() and Identifier.toString()
     }
 
     @Override
