--- a/net/minecraft/resources/Identifier.java
+++ b/net/minecraft/resources/Identifier.java
@@ -26,6 +_,10 @@
     public static final String PAPER_NAMESPACE = "paper"; // Paper
     private final String namespace;
     private final String path;
+    // Gale start - Cache - Identifier.hashCode() and Identifier.toString()
+    private @Nullable String cachedString;
+    private int cachedHashCode;
+    // Gale end - Cache - Identifier.hashCode() and Identifier.toString()
 
     private Identifier(final String namespace, final String path) {
         assert isValidNamespace(namespace);
@@ -50,7 +_,11 @@
     }
 
     public static Identifier parse(final String identifier) {
-        return bySeparator(identifier, ':');
+        // Gale start - Cache - Identifier.hashCode() and Identifier.toString()
+        Identifier toReturn = bySeparator(identifier, ':');
+        toReturn.cachedString = identifier;
+        return toReturn;
+        // Gale end - Cache - Identifier.hashCode() and Identifier.toString()
     }
 
     public static Identifier withDefaultNamespace(final String path) {
@@ -131,7 +_,14 @@
 
     @Override
     public String toString() {
-        return this.namespace + ":" + this.path;
+        // Gale start - Cache - Identifier.hashCode() and Identifier.toString()
+        String string = this.cachedString;
+        if (string == null) {
+            string = this.namespace + ":" + this.path;
+            this.cachedString = string;
+        }
+        return string;
+        // Gale end - Cache - Identifier.hashCode() and Identifier.toString()
     }
 
     @Override
@@ -141,7 +_,14 @@
 
     @Override
     public int hashCode() {
-        return 31 * this.namespace.hashCode() + this.path.hashCode();
+        // Gale start - Cache - Identifier.hashCode() and Identifier.toString()
+        int hashCode = this.cachedHashCode;
+        if (hashCode == 0) {
+            hashCode = 31 * this.namespace.hashCode() + this.path.hashCode();
+            this.cachedHashCode = hashCode;
+        }
+        return hashCode;
+        // Gale end - Cache - Identifier.hashCode() and Identifier.toString()
     }
 
     @Override
