From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Tue, 29 Nov 2022 01:17:03 +0100
Subject: [PATCH] Hide irrelevant compilation warnings

License: GPL-3.0 (https://www.gnu.org/licenses/gpl-3.0.html)

diff --git a/build.gradle.kts b/build.gradle.kts
index 3103e811e1f41b782a5b95537b921c59cac2054f..c2f60f3f638d8a94046d20ff1dfca316a3395861 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -56,6 +56,16 @@ dependencies {
 }
 
 val craftbukkitPackageVersion = "1_19_R1" // Paper
+
+// Gale start - hide irrelevant compilation warnings
+tasks.withType<JavaCompile> {
+    val compilerArgs = options.compilerArgs
+    compilerArgs.add("-Xlint:-module")
+    compilerArgs.add("-Xlint:-removal")
+    compilerArgs.add("-Xlint:-dep-ann")
+}
+// Gale end - hide irrelevant compilation warnings
+
 tasks.jar {
     archiveClassifier.set("dev")
 
