From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Tue, 21 Mar 2023 17:56:47 +0100
Subject: [PATCH] fixup load portal destination chunk before entity teleport


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index b6404bfac381352b6ceb23a1b95ed5f3ccabe606..697fad59a6ab4fb41aac65979b42d8ed4b4c8c21 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -3486,7 +3486,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
                 if (entity != null) {
                     // Gale start - MultiPaper - load portal destination chunk before entity teleport
                     if (entity.level.galeConfig().gameplayMechanics.technical.loadPortalDestinationChunkBeforeEntityTeleport) {
-                        BlockPos pos = new BlockPos(position);
+                        BlockPos pos = BlockPos.containing(position);
                         worldserver.getChunkSource().addRegionTicket(TicketType.PORTAL, new ChunkPos(pos), 1, pos);
                         worldserver.getChunkAt(pos);
                     }
