From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Tue, 21 Mar 2023 17:59:14 +0100
Subject: [PATCH] fixup remove profiler


diff --git a/src/main/java/net/minecraft/world/entity/animal/sniffer/Sniffer.java b/src/main/java/net/minecraft/world/entity/animal/sniffer/Sniffer.java
index c6949c7ec0766bda07324a6f004b29cbd37d0f1f..b7e248b9589a52f561bddacdbdf873cdfec290b4 100644
--- a/src/main/java/net/minecraft/world/entity/animal/sniffer/Sniffer.java
+++ b/src/main/java/net/minecraft/world/entity/animal/sniffer/Sniffer.java
@@ -435,11 +435,8 @@ public class Sniffer extends Animal {
 
     @Override
     protected void customServerAiStep() {
-        this.level.getProfiler().push("snifferBrain");
         this.getBrain().tick((ServerLevel) this.level, this);
-        this.level.getProfiler().popPush("snifferActivityUpdate");
         SnifferAi.updateActivity(this);
-        this.level.getProfiler().pop();
         super.customServerAiStep();
     }
 
