--- a/net/minecraft/network/syncher/SynchedEntityData.java
+++ b/net/minecraft/network/syncher/SynchedEntityData.java
@@ -84,7 +_,15 @@
         }
 
         this.isDirty = false;
-        List<SynchedEntityData.DataValue<?>> result = new ArrayList<>();
+        // Gale start - Collections - SynchedEntityData.packDirty()
+        int count = 0;
+        for (SynchedEntityData.DataItem<?> dataItem : this.itemsById) {
+            if (dataItem.isDirty()) {
+                count++;
+            }
+        }
+        ArrayList<SynchedEntityData.DataValue<?>> result = new ArrayList<>(count);
+        // Gale end - Collections - SynchedEntityData.packDirty()
 
         for (SynchedEntityData.DataItem<?> dataItem : this.itemsById) {
             if (dataItem.isDirty()) {
