--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -102,6 +_,7 @@
     // Paper start - Folia shedulers
     public final io.papermc.paper.threadedregions.EntityScheduler taskScheduler = new io.papermc.paper.threadedregions.EntityScheduler(this);
     private final io.papermc.paper.threadedregions.scheduler.FoliaEntityScheduler apiScheduler = new io.papermc.paper.threadedregions.scheduler.FoliaEntityScheduler(this);
+    private org.galemc.gale.util.collection.PackedBooleanArray packedVisibilityInversions; // Gale - Intrusive values - CraftPlayer.invertedVisibilityEntities
 
     @Override
     public final io.papermc.paper.threadedregions.scheduler.EntityScheduler getScheduler() {
@@ -113,6 +_,23 @@
         this.server = server;
         this.entity = entity;
         this.entityType = CraftEntityType.minecraftToBukkit(entity.getType());
+        // Gale start - Intrusive values - CraftPlayer.invertedVisibilityEntities
+        UUID uuid = entity.getUUID();
+        if (CraftPlayer.occurrencesInInvertedVisibilityEntities.containsKey(uuid)) {
+            List<CraftPlayer> onlinePlayers = (List<CraftPlayer>) org.bukkit.Bukkit.getServer().getOnlinePlayers();
+            for (CraftPlayer player : onlinePlayers) {
+                if (player.invertedVisibilityEntities.containsKey(uuid)) {
+                    if (this.packedVisibilityInversions == null) {
+                        this.packedVisibilityInversions = org.galemc.gale.util.collection.PackedBooleanArray.createWithInitial(player.packedVisibilityInversionsWordIndex, player.packedVisibilityInversionsWordMask);
+                    } else {
+                        this.packedVisibilityInversions.set(player.packedVisibilityInversionsWordIndex, player.packedVisibilityInversionsWordMask);
+                    }
+                }
+            }
+        } else {
+            this.packedVisibilityInversions = org.galemc.gale.util.collection.PackedBooleanArray.createEmpty();
+        }
+        // Gale end - Intrusive values - CraftPlayer.invertedVisibilityEntities
     }
 
     public static <T extends Entity> CraftEntity getEntity(CraftServer server, T entity) {
@@ -578,7 +_,7 @@
     @Override
     public void setTicksLived(int value) {
         Preconditions.checkArgument(value > 0, "Age value (%s) must be greater than 0", value);
-        this.getHandle().tickCount = value;
+        this.getHandle().setTickCount(value); // Gale - Event-driven - Cat.canRemoveWhenFarAway, Ocelot.canRemoveWhenFarAway
         this.getHandle().totalEntityAge = value;
     }
 
@@ -736,6 +_,20 @@
         return this.getHandle().visibleByDefault;
     }
 
+    // Gale start - Intrusive values - CraftPlayer.invertedVisibilityEntities
+    public boolean isVisibilityInvertedFor(CraftPlayer player) {
+        return this.packedVisibilityInversions.get(player.packedVisibilityInversionsWordIndex, player.packedVisibilityInversionsWordMask);
+    }
+
+    public void setVisibilityInverted(CraftPlayer player) {
+        this.packedVisibilityInversions.set(player.packedVisibilityInversionsWordIndex, player.packedVisibilityInversionsWordMask);
+    }
+
+    public void unsetVisibilityInverted(CraftPlayer player) {
+        this.packedVisibilityInversions.unsetUnsafe(player.packedVisibilityInversionsWordIndex, player.packedVisibilityInversionsWordMask);
+    }
+    // Gale end - Intrusive values - CraftPlayer.invertedVisibilityEntities
+
     @Override
     public Set<Player> getTrackedBy() {
         Preconditions.checkState(!this.entity.generation, "Cannot get tracking players during world generation");
@@ -745,7 +_,7 @@
         ChunkMap.TrackedEntity entityTracker = world.getChunkSource().chunkMap.entityMap.get(this.getEntityId());
 
         if (entityTracker != null) {
-            for (ServerPlayerConnection connection : entityTracker.seenBy) {
+            for (ServerPlayerConnection connection : entityTracker.seenByList) { // Gale - Collections - ChunkMap.TrackedEntity.seenBy
                 players.add(connection.getPlayer().getBukkitEntity());
             }
         }
@@ -762,7 +_,10 @@
         ChunkMap.TrackedEntity entityTracker = world.getChunkSource().chunkMap.entityMap.get(this.getEntityId());
         if (entityTracker == null) return false;
 
-        return entityTracker.seenBy.contains(((CraftPlayer) player).getHandle().connection);
+        // Gale start - Collections - ChunkMap.TrackedEntity.seenBy
+        net.minecraft.server.network.ServerGamePacketListenerImpl connection = ((CraftPlayer) player).getHandle().connection;
+        return entityTracker.seenByPackedArray.get(connection.packedSeenByWordIndex, connection.packedSeenByWordMask);
+        // Gale end - Collections - ChunkMap.TrackedEntity.seenBy
     }
 
     @Override
@@ -1080,7 +_,7 @@
             return;
         }
 
-        for (final ServerPlayerConnection connection : entityTracker.seenBy) {
+        for (final ServerPlayerConnection connection : entityTracker.seenByList) { // Gale - Collections - ChunkMap.TrackedEntity.seenBy
             this.getHandle().resendPossiblyDesyncedEntityData(connection.getPlayer());
         }
     }
@@ -1225,8 +_,8 @@
             return java.util.Collections.emptySet();
         }
 
-        Set<org.bukkit.entity.Player> set = new java.util.HashSet<>(tracker.seenBy.size());
-        for (net.minecraft.server.network.ServerPlayerConnection connection : tracker.seenBy) {
+        Set<org.bukkit.entity.Player> set = new java.util.HashSet<>(tracker.seenByList.size()); // Gale - Collections - ChunkMap.TrackedEntity.seenBy
+        for (net.minecraft.server.network.ServerPlayerConnection connection : tracker.seenByList) { // Gale - Collections - ChunkMap.TrackedEntity.seenBy
             set.add(connection.getPlayer().getBukkitEntity().getPlayer());
         }
         return set;
