--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
@@ -43,6 +_,7 @@
     public void setDespawnInPeacefulOverride(final TriState state) {
         Preconditions.checkArgument(state != null, "TriState cannot be null");
         this.getHandle().despawnInPeacefulOverride = state;
+        this.getHandle().updateMustDespawnBecauseOfPeacefulDifficulty(); // Gale - Compute peaceful despawn reactively
     }
 
     @Override
