--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
@@ -42,7 +_,11 @@
     @Override
     public void setDespawnInPeacefulOverride(final TriState state) {
         Preconditions.checkArgument(state != null, "TriState cannot be null");
-        this.getHandle().despawnInPeacefulOverride = state;
+        // Gale start - Event-driven - Mob.shouldDespawnInPeaceful
+        net.minecraft.world.entity.Mob handle = this.getHandle();
+        handle.despawnInPeacefulOverride = state;
+        handle.gale$eventDriven_gale$shouldDespawnInPeaceful_update();
+        // Gale end - Event-driven - Mob.shouldDespawnInPeaceful
     }
 
     @Override
