--- a/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java
+++ b/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java
@@ -36,14 +_,21 @@
 
     // SimplePluginManager
     public void callEvent(@NotNull Event event) {
+        // Gale start - Do less work - Exit Bukkit event call early if no listeners
+        RegisteredListener[] listeners = event.getHandlers().getRegisteredListeners();
+        if (listeners.length == 0) return;
+        // Gale end - Do less work - Exit Bukkit event call early if no listeners
+
         if (event.isAsynchronous() && this.server.isPrimaryThread()) {
             throw new IllegalStateException(event.getEventName() + " may only be triggered asynchronously.");
         } else if (!event.isAsynchronous() && !this.server.isPrimaryThread() && !this.server.isStopping()) {
             throw new IllegalStateException(event.getEventName() + " may only be triggered synchronously.");
         }
 
-        HandlerList handlers = event.getHandlers();
-        RegisteredListener[] listeners = handlers.getRegisteredListeners();
+        // Gale start - Do less work - Exit Bukkit event call early if no listeners
+//        HandlerList handlers = event.getHandlers();
+//        RegisteredListener[] listeners = handlers.getRegisteredListeners();
+        // Gale end - Do less work - Exit Bukkit event call early if no listeners
 
         for (RegisteredListener registration : listeners) {
             if (!registration.getPlugin().isEnabled()) {
