--- a/net/minecraft/world/level/chunk/ChunkGenerator.java
+++ b/net/minecraft/world/level/chunk/ChunkGenerator.java
@@ -363,10 +_,14 @@
             try {
                 Registry<PlacedFeature> featureRegistry = level.registryAccess().lookupOrThrow(Registries.PLACED_FEATURE);
                 int generationSteps = Math.max(GenerationStep.Decoration.values().length, featureStepCount);
+                // Gale start - Reuse structure decoration data
+                boolean generateStructures = structureManager.shouldGenerateStructures();
+                it.unimi.dsi.fastutil.objects.@Nullable ObjectArrayList<StructureStart> structureStarts = generateStructures ? new it.unimi.dsi.fastutil.objects.ObjectArrayList<>() : null;
+                // Gale end - Reuse structure decoration data
 
                 for (int stepIndex = 0; stepIndex < generationSteps; stepIndex++) {
                     int index = 0;
-                    if (structureManager.shouldGenerateStructures()) {
+                    if (generateStructures) { // Gale - Reuse structure decoration data
                         for (Structure structure : structuresByStep.getOrDefault(stepIndex, Collections.emptyList())) {
                             random.setFeatureSeed(decorationSeed, index, stepIndex);
                             Supplier<String> currentlyGenerating = () -> structuresRegistry.getResourceKey(structure)
@@ -375,8 +_,16 @@
 
                             try {
                                 level.setCurrentlyGenerating(currentlyGenerating);
-                                structureManager.startsForStructure(sectionPos, structure)
-                                    .forEach(start -> start.placeInChunk(level, structureManager, this, random, getWritableArea(chunk), centerPos));
+                                // Gale start - Reuse structure decoration data
+                                structureStarts.clear();
+                                structureManager.fillStartsForStructure(structure, chunk.getReferencesForStructure(structure), structureStarts::add);
+                                Object[] structureStartsArray = structureStarts.elements();
+                                for (int structureStartIndex = 0, structureStartsSize = structureStarts.size(); structureStartIndex < structureStartsSize; structureStartIndex++) {
+                                    ((StructureStart) structureStartsArray[structureStartIndex]).placeInChunk(level, structureManager, this, random, getWritableArea(chunk), centerPos);
+                                }
+                                // Gale end - Reuse structure decoration data
                             } catch (Exception e) {
                                 CrashReport report = CrashReport.forThrowable(e, "Feature placement");
                                 report.addCategory("Feature").setDetail("Description", currentlyGenerating::get);
