--- a/net/minecraft/world/entity/ai/navigation/PathNavigation.java
+++ b/net/minecraft/world/entity/ai/navigation/PathNavigation.java
@@ -411,7 +_,7 @@
                 Node node = this.path.getNode(i);
                 Node nextNode = i + 1 < this.path.getNodeCount() ? this.path.getNode(i + 1) : null;
                 BlockState state = this.level.getBlockState(new BlockPos(node.x, node.y, node.z));
-                if (state.is(BlockTags.CAULDRONS)) {
+                if (state.isAny(org.dreeam.leaf.util.BlockMasks.CAULDRONS_TAG)) { // Gale - Pre-compute - Common BlockState predicates - Apply
                     this.path.replaceNode(i, node.cloneAndMove(node.x, node.y + 1, node.z));
                     if (nextNode != null && node.y >= nextNode.y) {
                         this.path.replaceNode(i + 1, node.cloneAndMove(nextNode.x, node.y + 1, nextNode.z));
